<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 10 Intro to Genomic Prediction | Genomic Prediction and Selection Manual</title>
<meta name="author" content="Marnin Wolfe">
<meta name="description" content="In previous sections, we processed and prepared genotype data for downstream analyses. We also gathered and prepared phenotypic data, ultimately summarizing the plot-basis data into BLUP values...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 10 Intro to Genomic Prediction | Genomic Prediction and Selection Manual">
<meta property="og:type" content="book">
<meta property="og:url" content="https://wolfemd.github.io/GenomicSelectionManual/intro-to-genomic-prediction.html">
<meta property="og:description" content="In previous sections, we processed and prepared genotype data for downstream analyses. We also gathered and prepared phenotypic data, ultimately summarizing the plot-basis data into BLUP values...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 10 Intro to Genomic Prediction | Genomic Prediction and Selection Manual">
<meta name="twitter:description" content="In previous sections, we processed and prepared genotype data for downstream analyses. We also gathered and prepared phenotypic data, ultimately summarizing the plot-basis data into BLUP values...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="libs/pagedtable-1.1/js/pagedtable.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Genomic Prediction and Selection Manual</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Preamble</a></li>
<li><a class="" href="overview-process-map.html"><span class="header-section-number">2</span> Overview Process Map</a></li>
<li><a class="" href="using-this-manual.html"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="" href="process-maps.html"><span class="header-section-number">4</span> Process maps</a></li>
<li><a class="" href="create_project.html"><span class="header-section-number">5</span> Create a project</a></li>
<li><a class="" href="download-training-data.html"><span class="header-section-number">6</span> Download training data</a></li>
<li><a class="" href="prepare-phenotype-data.html"><span class="header-section-number">7</span> Prepare phenotype data</a></li>
<li><a class="" href="prepare-genotypic-data.html"><span class="header-section-number">8</span> Prepare genotypic data</a></li>
<li><a class="" href="preliminary-field-trial-analysis.html"><span class="header-section-number">9</span> Preliminary field trial analysis</a></li>
<li><a class="active" href="intro-to-genomic-prediction.html"><span class="header-section-number">10</span> Intro to Genomic Prediction</a></li>
<li><a class="" href="standard-k-fold-cross-validation.html"><span class="header-section-number">11</span> Standard K-fold Cross-validation</a></li>
<li><a class="" href="predict-parental-breeding-values.html"><span class="header-section-number">12</span> Predict parental breeding values</a></li>
<li><a class="" href="intro-to-genomic-cross-prediction.html"><span class="header-section-number">13</span> Intro to Genomic Cross Prediction</a></li>
<li><a class="" href="parentwise_cross_val.html"><span class="header-section-number">14</span> Accuracy of cross prediction?</a></li>
<li><a class="" href="predict-crosses.html"><span class="header-section-number">15</span> Predict crosses</a></li>
<li><a class="" href="references.html">References</a></li>
<li><a class="" href="workflowr-example.html"><span class="header-section-number">16</span> WorkflowR example</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/wolfemd/GenomicSelectionManual">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="intro-to-genomic-prediction" class="section level1" number="10">
<h1>
<span class="header-section-number">10</span> Intro to Genomic Prediction<a class="anchor" aria-label="anchor" href="#intro-to-genomic-prediction"><i class="fas fa-link"></i></a>
</h1>
<p>In previous sections, we processed and <a href="prepare-genotypic-data">prepared genotype data</a> for downstream analyses. We also gathered and <a href="prepare-phenotype-data">prepared phenotypic data</a>, ultimately summarizing the plot-basis data into <a href="preliminary-field-trial-analysis">BLUP values for each individual</a> in out training population.</p>
<p>We have the data ready to do genomic predictions.</p>
<p>Our process map indicates that the next step is to use cross-validation to check prediction accuracy, before completing prediction of <strong>genomic estimated breeding values (GEBV)</strong> for selections.</p>
<p>Before proceeding with the applied pipeline, work through this section and an introduction to <strong>genomic prediction in R</strong>.</p>
<p>If you have not had any training in quantitative genetics, mixed-models or genomic prediction, consider jumping to the bottom section of this chapter <a href="intro-to-genomic-prediction.html#recommended-literature">recommended literature</a> for learning resources.</p>
<div id="librarygenomicmateselectr" class="section level2" number="10.1">
<h2>
<span class="header-section-number">10.1</span> <code>library(genomicMateSelectR)</code><a class="anchor" aria-label="anchor" href="#librarygenomicmateselectr"><i class="fas fa-link"></i></a>
</h2>
<p>To facilitate the genomic selection pipeline, I built an R package, <code>genomicMateSelectR</code>, which can do the cross-validation, predictions and more using only a few functions. We already encountered a few of the functions in the package previously, when we <a href="filter-variants">filtered variants</a>, <a href="construct-grms">constructed kinship matrices</a> and also a <a href="recomb-freq-mat">recombination frequency matrix</a>.</p>
<p>In those previous steps, we focused on simply applying the functions and creating the inputs that the process map indicates we will need. We didn’t spend much time previously discussing what these inputs were or why we wanted them.</p>
<p>The functions from <code>genomicMateSelectR</code>, are themselves using functions from <code><a href="https://rdrr.io/r/base/library.html">library(sommer)</a></code>. Rather than simply applying the functions as instructed, we’ll dig under-the-hood to understand what is happening.</p>
<p>It is <em>also</em> recommended that you check-out and work through the <a href="https://wolfemd.github.io/genomicMateSelectR/articles/non_additive_models.html">vignettes for <code>genomicMateSelectR</code></a>, which should also help.</p>
<p>At the bottom of this chapter, I’ll provide links and references for learning more. I <em>highly</em> recommend reading and learning from multiple sources, as no single author explains everything perfectly for everyone.</p>
</div>
<div id="genomic-blup" class="section level2" number="10.2">
<h2>
<span class="header-section-number">10.2</span> Genomic BLUP<a class="anchor" aria-label="anchor" href="#genomic-blup"><i class="fas fa-link"></i></a>
</h2>
<p>In the <a href="preliminary-field-trial-analysis">previous section</a>, we learned to use the <code>mmer()</code> function in <code><a href="https://rdrr.io/r/base/library.html">library(sommer)</a></code> to fit mixed-models <em>without</em> including genomic information.</p>
<p>We modeled the <strong>germplasmName</strong> variable, which distinguishes unique genotypes (i.e. cassava clones) as a random-effect and extracted the predicted <strong>BLUP</strong> for each clone as a measure of performance. In that model, each unique <strong>germpasmName</strong> was modeled by default as <em>independent and identically distributed</em> (<em>i.i.d.</em>). In other words, lacking additional information to input into the model, we assumed that each clone is unrelated to each other clone. In such a situation, we can only obtain performance predictions for clones that had phenotypes.</p>
<p>Here is a simple, mixed-model notation, ignoring fixed-effects for simplicity:</p>
<p><span class="math display">\[\boldsymbol{y} =\mu + \boldsymbol{g} +\boldsymbol{\epsilon}\]</span> The first equation says the phenotype <span class="math inline">\(y\)</span> is predicted by a overall population mean <span class="math inline">\(\mu\)</span> + the genetic values of each individual <strong>germplasmName</strong> (<span class="math inline">\(g\)</span>, basically the genotype or clone mean) plus some unexplained residual values (<span class="math inline">\(\epsilon\)</span>).</p>
<p><span class="math display">\[g \sim N(0,I\sigma^2_{g})\]</span></p>
<p>This second notation says that the term <span class="math inline">\(g\)</span> is modeled as a random effect, drawn from a normal distribution (indicated by the N) that has a mean of zero, and a variance parameter <span class="math inline">\(\sigma^2_{g}\)</span> and a covariance <span class="math inline">\(\boldsymbol{I}\)</span>. Since <span class="math inline">\(\boldsymbol{I}\)</span> stands for the identity matrix, which has zeros off-diagonal, and 1’s on the diagonal, all clones are to be modeled as being unrelated; in other words, <em>independent and identically distributed</em> (<em>i.i.d.</em>).</p>
<p>We know that this assumption is incorrect. All of the individuals in our population are expected to be related at some level.</p>
<p>To make this a genomic prediction model, we will primarily use a mixed-model approach known as “genomic BLUP” or <strong>GBLUP</strong> for short.</p>
<p>In the <strong>GBLUP</strong> model, instead of using an identity matrix <span class="math inline">\(I\)</span> to say that all clones are unrelated, we will directly measure the degree of relatedness of our individuals using SNP markers distributed genome-wide. We will create what is called a “genomic relatedness matrix” (GRM), also called often called a kinship matrix and often notated as <span class="math inline">\(\boldsymbol{K}\)</span>.</p>
<p>So to make our mixed-model used in the previous section to get BLUPs into a genomic prediction model, we “simply” replace the <span class="math inline">\(\boldsymbol{I}\)</span> with<span class="math inline">\(\boldsymbol{K}\)</span> in the mixed-model equations:</p>
<p><span class="math display">\[g \sim N(0,K\sigma^2_{g})\]</span> <em>We are assuming the relative similarity or dissimilarity between the performance of our clones is proportional to their degree of relatedness.</em></p>
<p>We model the <strong>germplasmName</strong> variable with a covariance, the GRM in our mixed-model and obtain BLUPs, which we can call GBLUPs, that are conditioned on the resemblance between relatives. As long as the GRM is constructed using a particular formula, we will be predicted breeding values, and thus the term you’ve all heard: <em>genomic estimated breeding values</em> (<strong>GEBVs</strong>).</p>
<p><em>The mixed-model equations can accommodate prediction of individuals that have genotypes and are in the kinship matrix, even if they do not have phenotypes.</em> In other words, we can predict the performance of individuals on the basis of their (ideally close) relatives.</p>
<div id="getting-started" class="section level3" number="10.2.1">
<h3>
<span class="header-section-number">10.2.1</span> Getting started<a class="anchor" aria-label="anchor" href="#getting-started"><i class="fas fa-link"></i></a>
</h3>
<p>Start with BLUPs from the <a href="preliminary-field-trial-analysis">previous step</a>.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="kinship-matrix" class="section level3" number="10.2.2">
<h3>
<span class="header-section-number">10.2.2</span> Kinship matrix<a class="anchor" aria-label="anchor" href="#kinship-matrix"><i class="fas fa-link"></i></a>
</h3>
<p>In the chapter on <a href="prepare-genotypic-data">preparing genotypic data</a>, we used database-sourced genotype data and the <code>genomicMateSelectR</code> function <code><a href="https://rdrr.io/pkg/genomicMateSelectR/man/kinship.html">kinship()</a></code> to create an additive genomic relationship matrix.</p>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">A</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"kinship_add.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Let’s take a look at the contents of the matrix:</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">A</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;                    IITA-TMS-IBA30572 IITA-TMS-IBA940237</span>
<span class="co">#&gt; IITA-TMS-IBA30572         0.86194941        -0.09134081</span>
<span class="co">#&gt; IITA-TMS-IBA940237       -0.09134081         0.95954799</span>
<span class="co">#&gt; IITA-TMS-IBA961642       -0.20180585         0.04458302</span>
<span class="co">#&gt; IITA-TMS-ONN920168        0.15812456        -0.09358516</span>
<span class="co">#&gt; IITA-TMS-WAR4080          0.82642409        -0.09229778</span>
<span class="co">#&gt;                    IITA-TMS-IBA961642 IITA-TMS-ONN920168</span>
<span class="co">#&gt; IITA-TMS-IBA30572         -0.20180585         0.15812456</span>
<span class="co">#&gt; IITA-TMS-IBA940237         0.04458302        -0.09358516</span>
<span class="co">#&gt; IITA-TMS-IBA961642         1.00259446        -0.16869620</span>
<span class="co">#&gt; IITA-TMS-ONN920168        -0.16869620         0.95173805</span>
<span class="co">#&gt; IITA-TMS-WAR4080          -0.19569202         0.15481065</span>
<span class="co">#&gt;                    IITA-TMS-WAR4080</span>
<span class="co">#&gt; IITA-TMS-IBA30572        0.82642409</span>
<span class="co">#&gt; IITA-TMS-IBA940237      -0.09229778</span>
<span class="co">#&gt; IITA-TMS-IBA961642      -0.19569202</span>
<span class="co">#&gt; IITA-TMS-ONN920168       0.15481065</span>
<span class="co">#&gt; IITA-TMS-WAR4080         0.85217903</span></code></pre></div>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span>
<span class="co">#&gt; [1] 963 963</span></code></pre></div>
<p>Notice that it is square (same number rows and columns) and symmetrix around the diagonal values.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">A</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html">lower.tri</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">]</span>, main<span class="op">=</span><span class="st">'Off-diagonal values (lower-triangle) of the kinship matrix'</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="06-intro_to_gp_files/figure-html/unnamed-chunk-5-1.png" width="672"></div>
<p>The off-diagonals express the genomic relatedness between different individuals in the dataset.</p>
<p>This formulation has a mean-value of zero, such that &gt;0 kinship means “more than the average population level of relatedness” while &lt;0 values mean “greater than average.”</p>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span>, main<span class="op">=</span><span class="st">'Diagonals of the kinship'</span>, breaks<span class="op">=</span><span class="fl">20</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="06-intro_to_gp_files/figure-html/unnamed-chunk-6-1.png" width="672"></div>
<p>The diagonal values express the relatedness of each individual to itself, which should be interpreted as the level of inbreeding (homozygosity).</p>
<p>The matrix is constructed using the same formula implemented in the “classic” <code>rrBLUP</code> function <code>A.mat()</code>: From <span class="citation">VanRaden (2008)</span>, Method 1.</p>
<p>I find the paper by Vitezica et al. 2013 particularly helpful in understanding additive (and non-additive) genomic relationships: <span class="citation">Vitezica <em>et al.</em> (2013)</span>.</p>
<p><strong>To understand more:</strong></p>
<ul>
<li><a href="http://nce.ads.uga.edu/wiki/lib/exe/fetch.php?media=uga_3_relationship.pdf" class="uri">http://nce.ads.uga.edu/wiki/lib/exe/fetch.php?media=uga_3_relationship.pdf</a></li>
<li><a href="https://colloque.inrae.fr/iufro2016/content/download/5576/73493/version/1/file/Legarra_IUFROArcachon_2016.pdf" class="uri">https://colloque.inrae.fr/iufro2016/content/download/5576/73493/version/1/file/Legarra_IUFROArcachon_2016.pdf</a></li>
<li><a href="https://plant-breeding-genomics.extension.org/genomic-relationships-and-gblup/" class="uri">https://plant-breeding-genomics.extension.org/genomic-relationships-and-gblup/</a></li>
</ul>
</div>
<div id="genomic-prediction" class="section level3" number="10.2.3">
<h3>
<span class="header-section-number">10.2.3</span> Genomic prediction<a class="anchor" aria-label="anchor" href="#genomic-prediction"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pull out one trait (DM) BLUPs</span>
<span class="va">dm_blups</span><span class="op">&lt;-</span><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="va">dm_blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="va">head</span>
<span class="co">#&gt; # A tibble: 6 × 6</span>
<span class="co">#&gt;   germplasmName        BLUP   PEV   REL drgBLUP    WT</span>
<span class="co">#&gt;   &lt;chr&gt;               &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 IITA-TMS-IBA000070 -0.776 0.172 0.963  -0.806  5.27</span>
<span class="co">#&gt; 2 IITA-TMS-IBA070593 -4.38  0.389 0.917  -4.78   3.82</span>
<span class="co">#&gt; 3 IITA-TMS-IBA30572  -0.690 0.162 0.965  -0.715  5.37</span>
<span class="co">#&gt; 4 IITA-TMS-IBA980581 -1.17  0.173 0.963  -1.21   5.26</span>
<span class="co">#&gt; 5 IITA-TMS-IBA982101 -1.94  0.467 0.900  -2.15   3.45</span>
<span class="co">#&gt; 6 TMEB419             1.59  0.240 0.949   1.68   4.73</span></code></pre></div>
<p>Now we have all we need to implement a GBLUP model using <code>mmer()</code>.</p>
<p>One thing to note:</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">dm_blups</span><span class="op">$</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;   463   346</span></code></pre></div>
<p>If you go back to the chapter on <a href="prepare-genotypic-data">preparing genotypic data</a> you will recall that in this dataset, there are phenotyped-but-not-genotyped lines.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">dm_blups</span><span class="op">$</span><span class="va">germplasmName</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;   617   346</span></code></pre></div>
<p>There are also genotyped-but-not-phenotyped lines.</p>
<p>It is always important to check this and to verify that the result you are getting makes sense. Are there accessions that are not appearing genotyped (or phenotyped), which you were expecting? If so, you need to trace it. There may be names that didn’t match, which could require fixing <em>preferably in the database to ensure long-term improvement in the ease and correctness of your analyses</em>.</p>
<p>Phenotyped-but-not-genotyped lines need to be excluded from the analysis. Genotyped-but-not-phenotyped lines, as discussed above are accepted and can be predicted.</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dm_blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This time, the de-regressed BLUPs (<strong>drgBLUP</strong>) from the previous step are the response data. We will use the weights (<strong>WT</strong>) computed previously in the <code>weights=</code> argument of <code>mmer()</code>.</p>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gblup_all</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="va">drgBLUP</span><span class="op">~</span><span class="fl">1</span>,
                <span class="co"># here we specify a random-effect for the "germplasmName" variable</span>
                <span class="co"># and supply the kinship matrix "A" as follows:</span>
                random <span class="op">=</span> <span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="va">germplasmName</span>,Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span>,
                weights <span class="op">=</span> <span class="va">WT</span>,
                data<span class="op">=</span><span class="va">dm_blups</span><span class="op">)</span>
<span class="co">#&gt; Adding additional levels of Gu in the model matrix of 'germplasmName' </span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -187.325   16:33:10      0           0</span>
<span class="co">#&gt;     2      -187.167   16:33:10      0           0</span>
<span class="co">#&gt;     3      -187.095   16:33:11      1           0</span>
<span class="co">#&gt;     4      -187.077   16:33:11      1           0</span>
<span class="co">#&gt;     5      -187.075   16:33:11      1           0</span>
<span class="co">#&gt;     6      -187.075   16:33:11      1           0</span></code></pre></div>
<p>Here’s how to get out the <strong>GEBV</strong> from the <code>mmer()</code> output:</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gebv</span><span class="op">&lt;-</span><span class="va">gblup_all</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:germplasmName`</span><span class="op">$</span><span class="va">drgBLUP</span>
<span class="co"># Notice the "germplasmName" and "drgBLUP" references in this call </span>
<span class="co">## are specific to the analysis/dataset in question. </span>
<span class="co">## See the sommer manual on CRAN. </span></code></pre></div>
<p>Things to notice:</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">gebv</span><span class="op">)</span>
<span class="co">#&gt; [1] 963</span></code></pre></div>
<p>Only 346 phenotyped lines were included. 963 genotyped lines in the kinship matrix. 963 GEBV predicted!</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">gebv</span><span class="op">)</span>
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt; -2.9475 -0.3902  0.0270  0.0000  0.4331  1.6265</span></code></pre></div>
<p>GEBV are centered on zero.</p>
</div>
<div id="prediction-accuracy" class="section level3" number="10.2.4">
<h3>
<span class="header-section-number">10.2.4</span> Prediction accuracy<a class="anchor" aria-label="anchor" href="#prediction-accuracy"><i class="fas fa-link"></i></a>
</h3>
<p>Now we can move into a more potentially complicated arena: evaluating prediction accuracy.</p>
<p>There are many scenarios and styles for doing this.</p>
<p>Let’s start <em>super</em> simple.</p>
<p>Of the 346 phenotyped lines, I randomly sample approx. 1/5 (20%) or 70 lines.</p>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1212</span><span class="op">)</span>
<span class="va">test_set</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">dm_blups</span><span class="op">$</span><span class="va">germplasmName</span>,size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span><span class="op">(</span><span class="fl">346</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>Remove the lines chosen (<strong>test_set</strong>) from the training dataset:</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">training_blups</span><span class="op">&lt;-</span><span class="va">dm_blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">test_set</span><span class="op">)</span></code></pre></div>
<p>Fit the prediction model:</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gblup_train</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="va">drgBLUP</span><span class="op">~</span><span class="fl">1</span>,
                  <span class="co"># here we specify a random-effect for the "germplasmName" variable</span>
                  <span class="co"># and supply the kinship matrix "A" as follows:</span>
                  random <span class="op">=</span> <span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="va">germplasmName</span>,Gu<span class="op">=</span><span class="va">A</span><span class="op">)</span>,
                  weights <span class="op">=</span> <span class="va">WT</span>,
                  data<span class="op">=</span><span class="va">training_blups</span><span class="op">)</span>
<span class="co">#&gt; Adding additional levels of Gu in the model matrix of 'germplasmName' </span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -151.68   16:33:12      0           0</span>
<span class="co">#&gt;     2      -151.327   16:33:12      0           0</span>
<span class="co">#&gt;     3      -151.132   16:33:12      0           0</span>
<span class="co">#&gt;     4      -151.062   16:33:12      0           0</span>
<span class="co">#&gt;     5      -151.049   16:33:12      0           0</span>
<span class="co">#&gt;     6      -151.046   16:33:12      0           0</span>
<span class="co">#&gt;     7      -151.046   16:33:12      0           0</span></code></pre></div>
<p>Again, all 963 lines in the kinship matrix get GEBV output.</p>
<p>Now we can compare (correlate) the GEBV predicted for the held-out <strong>test_set</strong> lines to the BLUPs for those lines.</p>
<p>This is considered an estimate of <strong>“prediction accuracy”</strong>.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># grab the gebv only for the test_set lines</span>
<span class="va">gebv_test</span><span class="op">&lt;-</span><span class="va">gblup_train</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:germplasmName`</span><span class="op">$</span><span class="va">drgBLUP</span><span class="op">[</span><span class="va">test_set</span><span class="op">]</span>
<span class="co"># merge the original BLUPs (for the test set only) to the corresponding GEBV</span>
<span class="co"># I use a left_join() or merge() to be absolutely sure there are no mix-ups</span>
<span class="va">gebv_vs_blups_testset</span><span class="op">&lt;-</span><span class="va">dm_blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">test_set</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>germplasmName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gebv_test</span><span class="op">)</span>,GEBV<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">gebv_test</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "germplasmName"</span></code></pre></div>
<p>Correlate the GEBV with the BLUP for the test_set.</p>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gebv_vs_blups_testset</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/exposition.html">%$%</a></span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">BLUP</span>,<span class="va">GEBV</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.09310568</span></code></pre></div>
<p>This correlation represents an estimate of accuracy predicting lines that aren’t phenotyped, based only on their genetic relatedness and the phenotypes of their relatives.</p>
<p>Sure, this is a low values. But it’s a small dataset using a small sample of markers to make compute fast for example purposes.</p>
</div>
</div>
<div id="cross-validation" class="section level2" number="10.3">
<h2>
<span class="header-section-number">10.3</span> Cross-validation<a class="anchor" aria-label="anchor" href="#cross-validation"><i class="fas fa-link"></i></a>
</h2>
<p>At it’s simplest, cross-validation involves assessing the expected accuracy of predicting untested lines by doing the above procedure multiple times across multiple random samples.</p>
<p>Here’s a nice graphic to illustrate:</p>
<div class="inline-figure"><img src="images/paste-63480391.png"></div>
<p><a href="https://imada.sdu.dk/~marco/Teaching/AY2010-2011/DM825/">Click here</a> for the original, animated version.</p>
<p>Do a quick google search for “k-fold cross validation” and you find plenty of good graphical explanations.</p>
<p>Cross-validation functions get complicated quickly as they have loops within loops. I built a function <code><a href="https://rdrr.io/pkg/genomicMateSelectR/man/runCrossVal.html">runCrossVal()</a></code> that fits several different models, handles multiple traits, can run in-parallel across mutiple compute-cores and even accepts selection index weights to compute selection index accuracy. <code><a href="https://rdrr.io/pkg/genomicMateSelectR/man/runCrossVal.html">runCrossVal()</a></code> is included in the <code>genomicMateSelectR</code> package.</p>
<p>Below, I use it on only 1 trait, additive-only model.</p>
<p><strong>NOTES:</strong> - doesn’t work properly <em>unless</em> <code>gid="GID"</code>. - It also requires you to remove any phenotyped lines that aren’t genotyped.</p>
<p>Following chunk does both of those:</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups_forRunCrossValFunc</span><span class="op">&lt;-</span><span class="va">blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># just one trait</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># need to rename the "blups" list to comply with the runCrossVal function</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>TrainingData<span class="op">=</span><span class="va">blups</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">TrainingData</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># need also to remove phenotyped-but-not-genotyped lines</span>
     <span class="co"># couldn't hurt to also subset the kinship to only phenotyped lines... would save RAM</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>TrainingData<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">TrainingData</span>,
                             <span class="op">~</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.</span>,<span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
                                  <span class="co"># rename the germplasmName column to GID</span>
                                  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>GID<span class="op">=</span><span class="va">germplasmName</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">standardCV</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/runCrossVal.html">runCrossVal</a></span><span class="op">(</span>blups<span class="op">=</span><span class="va">blups_forRunCrossValFunc</span>,
                        modelType<span class="op">=</span><span class="st">"A"</span>,
                        selInd<span class="op">=</span><span class="cn">FALSE</span>,
                        grms<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>A<span class="op">=</span><span class="va">A</span><span class="op">)</span>,
                        nrepeats<span class="op">=</span><span class="fl">2</span>,nfolds<span class="op">=</span><span class="fl">5</span>,
                        gid<span class="op">=</span><span class="st">"GID"</span>,seed<span class="op">=</span><span class="fl">424242</span>,
                        ncores<span class="op">=</span><span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: rsample</span>
<span class="co">#&gt; Loading required package: furrr</span>
<span class="co">#&gt; Loading required package: future</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -144.984   16:33:16      1           0</span>
<span class="co">#&gt;     2      -144.923   16:33:16      1           0</span>
<span class="co">#&gt;     3      -144.908   16:33:16      1           0</span>
<span class="co">#&gt;     4      -144.907   16:33:16      1           0</span>
<span class="co">#&gt;     5      -144.907   16:33:16      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -153.227   16:33:17      0           0</span>
<span class="co">#&gt;     2      -152.232   16:33:17      0           0</span>
<span class="co">#&gt;     3      -151.805   16:33:17      0           0</span>
<span class="co">#&gt;     4      -151.687   16:33:17      0           0</span>
<span class="co">#&gt;     5      -151.669   16:33:17      0           0</span>
<span class="co">#&gt;     6      -151.666   16:33:17      0           0</span>
<span class="co">#&gt;     7      -151.666   16:33:17      0           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -151.026   16:33:17      1           0</span>
<span class="co">#&gt;     2      -150.948   16:33:17      1           0</span>
<span class="co">#&gt;     3      -150.911   16:33:17      1           0</span>
<span class="co">#&gt;     4      -150.901   16:33:17      1           0</span>
<span class="co">#&gt;     5      -150.899   16:33:17      1           0</span>
<span class="co">#&gt;     6      -150.899   16:33:17      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -149.818   16:33:18      0           0</span>
<span class="co">#&gt;     2      -149.812   16:33:18      0           0</span>
<span class="co">#&gt;     3      -149.809   16:33:18      0           0</span>
<span class="co">#&gt;     4      -149.808   16:33:18      0           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -151.038   16:33:17      0           0</span>
<span class="co">#&gt;     2      -150.851   16:33:17      0           0</span>
<span class="co">#&gt;     3      -150.757   16:33:17      0           0</span>
<span class="co">#&gt;     4      -150.728   16:33:17      0           0</span>
<span class="co">#&gt;     5      -150.723   16:33:17      0           0</span>
<span class="co">#&gt;     6      -150.722   16:33:18      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -151.393   16:33:18      0           0</span>
<span class="co">#&gt;     2      -151.304   16:33:18      0           0</span>
<span class="co">#&gt;     3      -151.258   16:33:18      0           0</span>
<span class="co">#&gt;     4      -151.245   16:33:18      0           0</span>
<span class="co">#&gt;     5      -151.243   16:33:19      1           0</span>
<span class="co">#&gt;     6      -151.243   16:33:19      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -151.902   16:33:18      0           0</span>
<span class="co">#&gt;     2      -151.609   16:33:18      0           0</span>
<span class="co">#&gt;     3      -151.464   16:33:18      0           0</span>
<span class="co">#&gt;     4      -151.417   16:33:18      0           0</span>
<span class="co">#&gt;     5      -151.409   16:33:18      0           0</span>
<span class="co">#&gt;     6      -151.407   16:33:18      0           0</span>
<span class="co">#&gt;     7      -151.407   16:33:18      0           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -149.63   16:33:19      0           0</span>
<span class="co">#&gt;     2      -149.534   16:33:19      0           0</span>
<span class="co">#&gt;     3      -149.48   16:33:19      0           0</span>
<span class="co">#&gt;     4      -149.46   16:33:19      0           0</span>
<span class="co">#&gt;     5      -149.457   16:33:19      0           0</span>
<span class="co">#&gt;     6      -149.456   16:33:19      0           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -150.381   16:33:18      0           0</span>
<span class="co">#&gt;     2      -150.129   16:33:18      0           0</span>
<span class="co">#&gt;     3      -150.014   16:33:19      1           0</span>
<span class="co">#&gt;     4      -149.984   16:33:19      1           0</span>
<span class="co">#&gt;     5      -149.98   16:33:19      1           0</span>
<span class="co">#&gt;     6      -149.98   16:33:19      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -144.924   16:33:19      0           0</span>
<span class="co">#&gt;     2      -144.405   16:33:19      0           0</span>
<span class="co">#&gt;     3      -144.256   16:33:20      1           0</span>
<span class="co">#&gt;     4      -144.235   16:33:20      1           0</span>
<span class="co">#&gt;     5      -144.234   16:33:20      1           0</span>
<span class="co">#&gt; [1] "GBLUP model complete - one trait"</span>
<span class="co">#&gt; [1] "Genomic predictions done for all traits in one repeat-fold"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="co">#&gt; Joining, by = "GID"</span>
<span class="va">standardCV</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">accuracyEstOut</span><span class="op">)</span> 
<span class="co">#&gt; # A tibble: 10 × 9</span>
<span class="co">#&gt;    repeats  seeds splits     id    Trait predOf predVSobs   </span>
<span class="co">#&gt;      &lt;int&gt;  &lt;int&gt; &lt;list&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;list&gt;      </span>
<span class="co">#&gt;  1       1 395601 &lt;split [2… Fold1 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  2       1 395601 &lt;split [2… Fold2 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  3       1 395601 &lt;split [2… Fold3 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  4       1 395601 &lt;split [2… Fold4 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  5       1 395601 &lt;split [2… Fold5 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  6       2 215870 &lt;split [2… Fold1 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  7       2 215870 &lt;split [2… Fold2 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  8       2 215870 &lt;split [2… Fold3 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt;  9       2 215870 &lt;split [2… Fold4 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt; 10       2 215870 &lt;split [2… Fold5 DM    GEBV   &lt;tibble [96…</span>
<span class="co">#&gt; # … with 2 more variables: Accuracy &lt;dbl&gt;,</span>
<span class="co">#&gt; #   NcompleteTestPairs &lt;dbl&gt;</span></code></pre></div>
<p><strong>NOTE:</strong> Ignore the <code>NcompleteTestPairs</code> variable. It’s broken, but unimportant and I haven’t have time yet to fix it. I checked carefully and the actual prediction accuracy is correct.</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run this code to prove it, if you are interested :)</span>
<span class="co"># just checking the number of paired GEBV-BLUPs per test-set:</span>
<span class="co"># standardCV %&gt;% </span>
<span class="co">#      unnest(accuracyEstOut) %&gt;% </span>
<span class="co">#      mutate(NcompleteTestPairs=map_dbl(predVSobs,~na.omit(.) %&gt;% nrow(.)))</span></code></pre></div>
</div>
<div id="rrblup-and-equivalency-to-gblup" class="section level2" number="10.4">
<h2>
<span class="header-section-number">10.4</span> rrBLUP and equivalency to GBLUP<a class="anchor" aria-label="anchor" href="#rrblup-and-equivalency-to-gblup"><i class="fas fa-link"></i></a>
</h2>
<div id="rr-blup-aka-snp-blup" class="section level3" number="10.4.1">
<h3>
<span class="header-section-number">10.4.1</span> RR-BLUP aka SNP-BLUP<a class="anchor" aria-label="anchor" href="#rr-blup-aka-snp-blup"><i class="fas fa-link"></i></a>
</h3>
<p>I want to briefly cover and demonstrate an alternative, and under the right circumstances, equivalent model for genomic prediction: ridge-regression BLUP, aka SNP-BLUP. The rrBLUP model is a genome-wide marker regression model. There are many styles of this, especially in the Bayesian arena.</p>
<p>In marker-regression models, SNPs are used directly as predictors in a regression model, instead of using them to create a relatedness matrix. Both RRBLUP and GBLUP use a mixed-model approach.</p>
<p>For GBLUP we had:</p>
<p><span class="math display">\[\boldsymbol{y} =\mu + \boldsymbol{g} +\boldsymbol{\epsilon}\]</span> The random-effect predictor is simply the clone identity (germplasmName) and the covariance between those identities is input as the kinship matrix.</p>
<p>For RR-BLUP, we have instead:</p>
<p><span class="math display">\[\boldsymbol{y} =\mu + \boldsymbol{Zu} +\boldsymbol{\epsilon}\]</span></p>
<p><span class="math display">\[u \sim N(0,I\sigma^2_{u})\]</span></p>
<p><span class="math inline">\(\boldsymbol{Z}\)</span> is matrix of dimension <strong>N-individuals</strong> by <strong>P-SNPs</strong>. The vector <span class="math inline">\(\boldsymbol{u}\)</span> has dimension <strong>P-SNP</strong> by 1 and contains the BLUPs, which in this case represent SNP-effect predictions (SNP-BLUPs), with mean-effect 0, drawn <em>i.i.d.</em> from a normal distribution with variance <span class="math inline">\(\sigma^2_{u}\)</span>. Note that the variance parameter here is the variance among SNP-effects, <em>not</em> the variance among the <strong>GEBV</strong> of clones.</p>
<p><strong>GEBV</strong> are obtainable from this model by: <span class="math inline">\(\boldsymbol{g} = \boldsymbol{Zu}\)</span>.</p>
<p>The equivalence of RRBLUP and GBLUP come when the matrix of predictors, <span class="math inline">\(\boldsymbol{Z}\)</span> is centered in the same way as in the genomic relationshp matrix.</p>
<p><a href="https://htmlpreview.github.io/?https://github.com/lfelipe-ferrao/lfelipe-ferrao.github.io/blob/master/class/quantGenetic/week12.html"><strong>Click here</strong></a> and go to section 1.2.1 “VanRaden’s first genomic relationship matrix” for an explanation of the formula used for the kinship matrix.</p>
<p>Below, I’ll demonstrate:</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the dosage matrix</span>
<span class="va">M</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data"</span>,<span class="st">"dosages.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># re-load and extract the blups for just 1 trait</span>
<span class="va">blups</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">dm_blups</span><span class="op">&lt;-</span><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co"># remove phenotyped-but-not-genoypted blups</span>
<span class="va">dm_blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span>
<span class="co"># remove genotyped-but-not-phenoypted (not strictly necessary)</span>
<span class="va">M</span><span class="op">&lt;-</span><span class="va">M</span><span class="op">[</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">dm_blups</span><span class="op">$</span><span class="va">germplasmName</span>,<span class="op">]</span>
<span class="co"># Create a kinship matrix "K"</span>
<span class="va">K</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/kinship.html">kinship</a></span><span class="op">(</span><span class="va">M</span>,<span class="st">"add"</span><span class="op">)</span>
<span class="co"># Now create a centered dosage matrix, to use as predictors for rr-BLUP</span>
<span class="va">Z</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/centerDosage.html">centerDosage</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></code></pre></div>
<p>Fit the GBLUP model</p>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gblup</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="va">drgBLUP</span><span class="op">~</span><span class="fl">1</span>,
            random <span class="op">=</span> <span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="va">germplasmName</span>,Gu <span class="op">=</span> <span class="va">K</span><span class="op">)</span>,
            weights <span class="op">=</span> <span class="va">WT</span>,
            data<span class="op">=</span><span class="va">dm_blups</span><span class="op">)</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -187.324   16:33:21      1           0</span>
<span class="co">#&gt;     2      -187.166   16:33:21      1           0</span>
<span class="co">#&gt;     3      -187.095   16:33:21      1           0</span>
<span class="co">#&gt;     4      -187.077   16:33:21      1           0</span>
<span class="co">#&gt;     5      -187.075   16:33:21      1           0</span>
<span class="co">#&gt;     6      -187.075   16:33:21      1           0</span></code></pre></div>
<p>Now fit the RRBLUP model</p>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># one catch</span>
<span class="co"># the rows of the centered dosage matrix must be matched</span>
<span class="co">## to the rows of data.frame with the blups</span>
<span class="co">## one way to do this is by creating and incidence matrix</span>
<span class="co">## use the model.matrix() function as follows:</span>
<span class="va">ZincMat</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">germplasmName</span>,levels<span class="op">=</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>,data<span class="op">=</span><span class="va">dm_blups</span><span class="op">)</span> <span class="co"># -1 because we don't want an intercept here</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ZincMat</span><span class="op">)</span> <span class="co"># [1] 346 346 </span>
<span class="co">#&gt; [1] 346 346</span>
<span class="co"># relates the rows of the BLUPs to the rows the marker matrix</span>

<span class="co"># now fit the RRBLUP model</span>
<span class="va">rrblup</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="va">drgBLUP</span><span class="op">~</span><span class="fl">1</span>,
             random <span class="op">=</span> <span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">ZincMat</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span><span class="va">Z</span><span class="op">)</span>,buildGu <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,
             weights <span class="op">=</span> <span class="va">WT</span>,
             data<span class="op">=</span><span class="va">dm_blups</span><span class="op">)</span>
<span class="co">#&gt; iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#&gt;     1      -269.814   16:33:24      2           0</span>
<span class="co">#&gt;     2      -195.427   16:33:24      2           0</span>
<span class="co">#&gt;     3      -190.328   16:33:24      2           0</span>
<span class="co">#&gt;     4      -187.801   16:33:24      2           0</span>
<span class="co">#&gt;     5      -187.201   16:33:24      2           0</span>
<span class="co">#&gt;     6      -187.093   16:33:24      2           0</span>
<span class="co">#&gt;     7      -187.077   16:33:24      2           0</span>
<span class="co">#&gt;     8      -187.075   16:33:24      2           0</span>
<span class="co">#&gt;     9      -187.075   16:33:25      3           0</span></code></pre></div>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract the SNP-BLUPs</span>
<span class="va">u</span><span class="op">&lt;-</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">rrblup</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:Z`</span><span class="op">$</span><span class="va">drgBLUP</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">u</span><span class="op">)</span>
<span class="co">#&gt;        V1            </span>
<span class="co">#&gt;  Min.   :-0.0163276  </span>
<span class="co">#&gt;  1st Qu.:-0.0024412  </span>
<span class="co">#&gt;  Median :-0.0001831  </span>
<span class="co">#&gt;  Mean   :-0.0002119  </span>
<span class="co">#&gt;  3rd Qu.: 0.0020687  </span>
<span class="co">#&gt;  Max.   : 0.0151979</span></code></pre></div>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">u</span>,breaks<span class="op">=</span><span class="fl">20</span>,main <span class="op">=</span> <span class="st">'Distribution of SNP effects'</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="06-intro_to_gp_files/figure-html/unnamed-chunk-28-1.png" width="672"></div>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract the G-BLUPs (GEBVs)</span>
<span class="va">g</span><span class="op">&lt;-</span><span class="va">gblup</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">`u:germplasmName`</span><span class="op">$</span><span class="va">drgBLUP</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate the GEBVs using the SNP-effects (SNP-BLUPs)</span>
<span class="va">g_rr</span><span class="op">&lt;-</span><span class="va">Z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>,<span class="op">]</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span><span class="va">u</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">g_rr</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">g_rr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span>
<span class="co">#&gt; IITA-TMS-IBA000070 IITA-TMS-IBA070593  IITA-TMS-IBA30572 </span>
<span class="co">#&gt;         -0.1260352         -2.0363146         -0.7000067 </span>
<span class="co">#&gt; IITA-TMS-IBA980581 IITA-TMS-IBA982101 </span>
<span class="co">#&gt;         -0.9159732         -1.5152420</span></code></pre></div>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">g</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; IITA-TMS-IBA000070 IITA-TMS-IBA070593  IITA-TMS-IBA30572 </span>
<span class="co">#&gt;         -0.1254919         -2.0334329         -0.6996188 </span>
<span class="co">#&gt; IITA-TMS-IBA980581 IITA-TMS-IBA982101 </span>
<span class="co">#&gt;         -0.9151958         -1.5134661</span></code></pre></div>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cor(g,g_rr) # [1,] 0.9999997</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">g</span>,y<span class="op">=</span><span class="va">g_rr</span>, main <span class="op">=</span> <span class="st">'Proof that GEBV from GBLUP and RRBLUP are the equal!'</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>a<span class="op">=</span><span class="fl">0</span>,b<span class="op">=</span><span class="fl">1</span>,col<span class="op">=</span><span class="st">'red'</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="06-intro_to_gp_files/figure-html/unnamed-chunk-33-1.png" width="672"></div>
</div>
</div>
<div id="recommended-literature" class="section level2" number="10.5">
<h2>
<span class="header-section-number">10.5</span> Recommended Literature<a class="anchor" aria-label="anchor" href="#recommended-literature"><i class="fas fa-link"></i></a>
</h2>
<p>Check out this open-source 3-section short-course:<br>
“<a href="https://lfelipe-ferrao.github.io/class/survey/">Survey of Breeding Tools and (Genomic Selection) Methods</a>” by Particio Munoz and Felipe Ferrao at the U. of Florida. They do their best in a very short space to cover (1) introductory quantitative genetic models, (2) Statistical learning and whole-genome regression models, esp. rrBLUP and GBLUP models, and (3) theoretical and practical aspects of genomic selection.</p>
<p>I haven’t looked at all the details, but they also offer an open-source plant breeding focused “<a href="https://lfelipe-ferrao.github.io/class/quantGenetic/">Quantitative Genetics</a>” course!</p>
<p>Below is a list borrowed credit Felipe Ferrão’s course:</p>
<div id="quantitative-genetics" class="section level3" number="10.5.1">
<h3>
<span class="header-section-number">10.5.1</span> Quantitative Genetics<a class="anchor" aria-label="anchor" href="#quantitative-genetics"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Bernardo, 2010. Breeding for Quantitative Traits in Plant Breeding. <a href="http://stemmapress.com/">Book link</a>
</li>
<li>Falconer and Mackay, 1996. Introduction to Quantitative Genetics. <a href="https://www.amazon.com/Introduction-Quantitative-Genetics-Douglas-Falconer/dp/0582243025">Book link</a>
</li>
<li>Lynch and Walsh, 1998. Genetics and Analysis of Quantitative Traits. <a href="https://www.amazon.com/Genetics-Analysis-Quantitative-Traits-Michael/dp/0878934812">Book link</a>
</li>
</ul>
</div>
<div id="genomic-selection" class="section level3" number="10.5.2">
<h3>
<span class="header-section-number">10.5.2</span> Genomic Selection<a class="anchor" aria-label="anchor" href="#genomic-selection"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Wolfe 2016. Genomic Selection: Prediction Methodology &amp; Practical Application. Lecture Slides (gSlides) presented to the “Regional Southeast Asian Cassava Breeder’s Network Training” in November 2016. <a href="https://docs.google.com/presentation/d/1lFioutQq58IYs9-ewB7c24xfEHoy7TD5qAxNvJvYif8/edit?usp=sharing">gSlides Link</a>
</li>
<li>Mrode, 2014. Linear Models for the Prediction of Animal Breeding Values. <a href="https://www.amazon.com/Linear-Models-Prediction-Animal-Breeding/dp/1845939816">Book link</a>
</li>
<li>Isik et al., 2017. Genetic Data Analysis for Plant and Animal Breeding. <a href="https://www.springer.com/gp/book/9783319551753">Book link</a>
</li>
</ul>
</div>
<div id="additional-free-learning-resources" class="section level3" number="10.5.3">
<h3>
<span class="header-section-number">10.5.3</span> Additional (free) learning resources<a class="anchor" aria-label="anchor" href="#additional-free-learning-resources"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Learn more about genomic prediction</strong></p>
<ul>
<li><p><a href="http://nitro.biosci.arizona.edu/workshops/Synbreed2013/Lectures/Lecture28.pdf">Slides by Bruce Walsh on BLUP and Genomic Selection</a></p></li>
<li><p><a href="http://nitro.biosci.arizona.edu/workshops/MM2012/MM-2012.html">Mixed Models for Quantitative Genetics by Bruce Walsh</a></p></li>
<li><p><a href="https://zdoc.pub/lecture-5-blup-best-linear-unbiased-predictors-of-genetic-va.html">Lecture 5: BLUP (Best Linear Unbiased Predictors) of genetic values - by Bruce Walsh</a></p></li>
<li><p><a href="https://www.youtube.com/playlist?list=PLZ0lafzH_UmfHsoePj0t7xQglKSOpzbDp">Excellence in Breeding - Basics of GS - 2021 YouTube Video Series with Eduardo Covarrubias</a></p></li>
</ul>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="preliminary-field-trial-analysis.html"><span class="header-section-number">9</span> Preliminary field trial analysis</a></div>
<div class="next"><a href="standard-k-fold-cross-validation.html"><span class="header-section-number">11</span> Standard K-fold Cross-validation</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#intro-to-genomic-prediction"><span class="header-section-number">10</span> Intro to Genomic Prediction</a></li>
<li><a class="nav-link" href="#librarygenomicmateselectr"><span class="header-section-number">10.1</span> library(genomicMateSelectR)</a></li>
<li>
<a class="nav-link" href="#genomic-blup"><span class="header-section-number">10.2</span> Genomic BLUP</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#getting-started"><span class="header-section-number">10.2.1</span> Getting started</a></li>
<li><a class="nav-link" href="#kinship-matrix"><span class="header-section-number">10.2.2</span> Kinship matrix</a></li>
<li><a class="nav-link" href="#genomic-prediction"><span class="header-section-number">10.2.3</span> Genomic prediction</a></li>
<li><a class="nav-link" href="#prediction-accuracy"><span class="header-section-number">10.2.4</span> Prediction accuracy</a></li>
</ul>
</li>
<li><a class="nav-link" href="#cross-validation"><span class="header-section-number">10.3</span> Cross-validation</a></li>
<li>
<a class="nav-link" href="#rrblup-and-equivalency-to-gblup"><span class="header-section-number">10.4</span> rrBLUP and equivalency to GBLUP</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#rr-blup-aka-snp-blup"><span class="header-section-number">10.4.1</span> RR-BLUP aka SNP-BLUP</a></li></ul>
</li>
<li>
<a class="nav-link" href="#recommended-literature"><span class="header-section-number">10.5</span> Recommended Literature</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#quantitative-genetics"><span class="header-section-number">10.5.1</span> Quantitative Genetics</a></li>
<li><a class="nav-link" href="#genomic-selection"><span class="header-section-number">10.5.2</span> Genomic Selection</a></li>
<li><a class="nav-link" href="#additional-free-learning-resources"><span class="header-section-number">10.5.3</span> Additional (free) learning resources</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/wolfemd/GenomicSelectionManual/blob/master/06-intro_to_gp.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/wolfemd/GenomicSelectionManual/edit/master/06-intro_to_gp.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Genomic Prediction and Selection Manual</strong>" was written by Marnin Wolfe. It was last built on 2022-03-31.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
