<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 9 Preliminary field trial analysis | Genomic Prediction and Selection Manual</title>
<meta name="author" content="Marnin Wolfe">
<meta name="description" content="Context and Purpose: Upstream: Section 8 - quality control steps for genotypic data Downstream: Genomic prediction and related analyses. Inputs: Expected outputs:  9.1 Process Map   9.2 One-stage...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 9 Preliminary field trial analysis | Genomic Prediction and Selection Manual">
<meta property="og:type" content="book">
<meta property="og:url" content="https://wolfemd.github.io/GenomicSelectionManual/preliminary-field-trial-analysis.html">
<meta property="og:description" content="Context and Purpose: Upstream: Section 8 - quality control steps for genotypic data Downstream: Genomic prediction and related analyses. Inputs: Expected outputs:  9.1 Process Map   9.2 One-stage...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 9 Preliminary field trial analysis | Genomic Prediction and Selection Manual">
<meta name="twitter:description" content="Context and Purpose: Upstream: Section 8 - quality control steps for genotypic data Downstream: Genomic prediction and related analyses. Inputs: Expected outputs:  9.1 Process Map   9.2 One-stage...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="libs/pagedtable-1.1/js/pagedtable.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Genomic Prediction and Selection Manual</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Preamble</a></li>
<li><a class="" href="overview-process-map.html"><span class="header-section-number">2</span> Overview Process Map</a></li>
<li><a class="" href="using-this-manual.html"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="" href="process-maps.html"><span class="header-section-number">4</span> Process maps</a></li>
<li><a class="" href="create_project.html"><span class="header-section-number">5</span> Create a project</a></li>
<li><a class="" href="download-training-data.html"><span class="header-section-number">6</span> Download training data</a></li>
<li><a class="" href="prepare-phenotype-data.html"><span class="header-section-number">7</span> Prepare phenotype data</a></li>
<li><a class="" href="prepare-genotypic-data.html"><span class="header-section-number">8</span> Prepare genotypic data</a></li>
<li><a class="active" href="preliminary-field-trial-analysis.html"><span class="header-section-number">9</span> Preliminary field trial analysis</a></li>
<li><a class="" href="intro-to-genomic-prediction.html"><span class="header-section-number">10</span> Intro to Genomic Prediction</a></li>
<li><a class="" href="standard-k-fold-cross-validation.html"><span class="header-section-number">11</span> Standard K-fold Cross-validation</a></li>
<li><a class="" href="predict-parental-breeding-values.html"><span class="header-section-number">12</span> Predict parental breeding values</a></li>
<li><a class="" href="intro-to-genomic-cross-prediction.html"><span class="header-section-number">13</span> Intro to Genomic Cross Prediction</a></li>
<li><a class="" href="parentwise_cross_val.html"><span class="header-section-number">14</span> Accuracy of cross prediction?</a></li>
<li><a class="" href="predict-crosses.html"><span class="header-section-number">15</span> Predict crosses</a></li>
<li><a class="" href="references.html">References</a></li>
<li><a class="" href="workflowr-example.html"><span class="header-section-number">16</span> WorkflowR example</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/wolfemd/GenomicSelectionManual">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="preliminary-field-trial-analysis" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Preliminary field trial analysis<a class="anchor" aria-label="anchor" href="#preliminary-field-trial-analysis"><i class="fas fa-link"></i></a>
</h1>
<ul>
<li><p><strong>Context and Purpose:</strong></p></li>
<li><p><strong>Upstream:</strong> Section <a href="prepare-genotypic-data.html#prepare-genotypic-data">8</a> - quality control steps for genotypic data</p></li>
<li><p><strong>Downstream:</strong> Genomic prediction and related analyses.</p></li>
<li><p><strong>Inputs:</strong></p></li>
<li><p><strong>Expected outputs:</strong></p></li>
</ul>
<div id="process-map-3" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Process Map<a class="anchor" aria-label="anchor" href="#process-map-3"><i class="fas fa-link"></i></a>
</h2>
<div class="inline-figure"><img src="images/get_blups_process_map.png" style="width:100.0%"></div>
</div>
<div id="one-stage-or-multi-stage" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> One-stage or multi-stage?<a class="anchor" aria-label="anchor" href="#one-stage-or-multi-stage"><i class="fas fa-link"></i></a>
</h2>
<p>We often have very large, multi-year, multi-location, multi-trial-type (<strong>MET</strong>) datasets we use to train genomic prediction models. The number of plots can be in the range of 10- or even 100,000 plots with many thousands of unique genotypes observed in unbalanced fashion across heterogenous experimental designs. All that to say, the computational burden and level of expertise required to execute genomic prediction analyses directly on these data is very great.</p>
<p>For the sake of semi-automation and computational efficiency, the standard genomic prediction pipeline I have implemented for NextGen Cassava, which I demonstrate below is a two-stage weighted genomic prediction:</p>
<p><strong>Stage 1. Get BLUPs</strong></p>
<ul>
<li>
<p>Conduct preliminary analysis of the trial data <em>without</em> genomic relatedeness / marker-information.</p>
<p>Identify the best-fitting model for the data and potentially curate the raw data, removing outliers.</p>
</li>
<li><p>Fit mixed-model to <strong>MET</strong> data. Extract <strong><code>BLUPs</code></strong>, <strong><code>PEVs</code></strong> and variance components (<strong><code>VarComps</code></strong>). Compute de-regressed BLUPs (<strong><code>drgBLUPs</code></strong>) and weights (<strong><code>WTS</code></strong>) for “Stage 2.”</p></li>
</ul>
<p><strong>Stage 2. Cross-validation and genomic prediction</strong></p>
<p>Conduct weighted cross-validation and genomic prediction analyses using <strong><code>de-regressed BLUPs</code></strong> (alternative <strong><code>BLUEs</code></strong>) as the response variable and the weights from Stage 1. This effectively reduces the number of “observations” in the analysis to the number of unique genotypes (i.e. clones, inbred lines, etc.), leading to lower computational burden.</p>
<p><strong>Note of advice for single-stage analyses:</strong> I strongly recommend conducting preliminary analysis of the trial data <em>without</em> genomic relatedeness / marker-information. Ensure the model will converge, residuals look acceptable and otherwise assess the best fitting model <em>before</em> committing to lengthy computationally intensive analyses. :)</p>
</div>
<div id="set-up-training-datasets" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> Set-up training datasets<a class="anchor" aria-label="anchor" href="#set-up-training-datasets"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenos</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"phenotypes_cleaned.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The pipeline version of this analysis will use the <code>TRUE/FALSE</code> values of <strong><code>CompleteBlocks</code></strong> and <strong><code>IncompleteBlocks</code></strong> (<a href="prepare-phenotype-data.html#detect_designs">Preliminary analysis of trial data</a> ).</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">CompleteBlocks</span>,<span class="va">IncompleteBlocks</span>,<span class="va">locationName</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/spread.html">spread</a></span><span class="op">(</span><span class="va">locationName</span>,<span class="va">n</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 × 4</span>
<span class="co">#&gt;   CompleteBlocks IncompleteBlocks Ibadan Ubiaja</span>
<span class="co">#&gt;   &lt;lgl&gt;          &lt;lgl&gt;             &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 FALSE          TRUE                777     NA</span>
<span class="co">#&gt; 2 TRUE           FALSE                NA    125</span>
<span class="co">#&gt; 3 TRUE           TRUE                407    414</span></code></pre></div>
<p>Convert the data to “long format” . Remove missing values. “Nest” the data by Trait.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">traits</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DM"</span>,<span class="st">"MCMDS"</span>,<span class="st">"logFYLD"</span>,<span class="st">"logDYLD"</span><span class="op">)</span>
<span class="va">phenos</span><span class="op">&lt;-</span><span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># Convert the data to "long format" </span>
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">traits</span><span class="op">)</span>, 
                  names_to <span class="op">=</span> <span class="st">"Trait"</span>, values_to <span class="op">=</span> <span class="st">"Value"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="co"># Remove missing values</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">Value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="co"># Nest the MultiEnvironmentTrial data by trait</span>
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">nest</a></span><span class="op">(</span>METdata<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">Trait</span><span class="op">)</span><span class="op">)</span>
<span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>N_plots<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dbl</a></span><span class="op">(</span><span class="va">METdata</span>,<span class="va">nrow</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4 × 3</span>
<span class="co">#&gt;   Trait   METdata               N_plots</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;list&gt;                  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 DM      &lt;tibble [1,448 × 35]&gt;    1448</span>
<span class="co">#&gt; 2 MCMDS   &lt;tibble [1,436 × 35]&gt;    1436</span>
<span class="co">#&gt; 3 logFYLD &lt;tibble [1,590 × 35]&gt;    1590</span>
<span class="co">#&gt; 4 logDYLD &lt;tibble [1,443 × 35]&gt;    1443</span></code></pre></div>
<p>Where previously there was one row per plot and a large number of columns, now things are simple and tidy.</p>
<p>One row per trait. The actual plot-basis data are now contained within <strong>METdata</strong>, a list-type column, each element containing a <code>tibble</code>.</p>
<p>To demonstrate, check the contents of row 1 of the <strong>METdata</strong> column:</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenos</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["studyYear"],"name":[1],"type":["int"],"align":["right"]},{"label":["programName"],"name":[2],"type":["chr"],"align":["left"]},{"label":["locationName"],"name":[3],"type":["chr"],"align":["left"]},{"label":["studyName"],"name":[4],"type":["chr"],"align":["left"]},{"label":["studyDesign"],"name":[5],"type":["chr"],"align":["left"]},{"label":["plotWidth"],"name":[6],"type":["int"],"align":["right"]},{"label":["plotLength"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["fieldSize"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["plantingDate"],"name":[9],"type":["chr"],"align":["left"]},{"label":["harvestDate"],"name":[10],"type":["chr"],"align":["left"]},{"label":["germplasmName"],"name":[11],"type":["chr"],"align":["left"]},{"label":["observationUnitDbId"],"name":[12],"type":["int"],"align":["right"]},{"label":["replicate"],"name":[13],"type":["int"],"align":["right"]},{"label":["blockNumber"],"name":[14],"type":["int"],"align":["right"]},{"label":["plotNumber"],"name":[15],"type":["int"],"align":["right"]},{"label":["rowNumber"],"name":[16],"type":["int"],"align":["right"]},{"label":["colNumber"],"name":[17],"type":["int"],"align":["right"]},{"label":["entryType"],"name":[18],"type":["chr"],"align":["left"]},{"label":["trialType"],"name":[19],"type":["chr"],"align":["left"]},{"label":["numberBlocks"],"name":[20],"type":["int"],"align":["right"]},{"label":["numberReps"],"name":[21],"type":["int"],"align":["right"]},{"label":["observationUnitName"],"name":[22],"type":["chr"],"align":["left"]},{"label":["CompleteBlocks"],"name":[23],"type":["lgl"],"align":["right"]},{"label":["IncompleteBlocks"],"name":[24],"type":["lgl"],"align":["right"]},{"label":["yearInLoc"],"name":[25],"type":["chr"],"align":["left"]},{"label":["trialInLocYr"],"name":[26],"type":["chr"],"align":["left"]},{"label":["repInTrial"],"name":[27],"type":["chr"],"align":["left"]},{"label":["blockInRep"],"name":[28],"type":["chr"],"align":["left"]},{"label":["NOHAV"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["plotArea"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["MaxNOHAV"],"name":[31],"type":["dbl"],"align":["right"]},{"label":["plantsPerPlot"],"name":[32],"type":["dbl"],"align":["right"]},{"label":["PlantSpacing"],"name":[33],"type":["dbl"],"align":["right"]},{"label":["PropNOHAV"],"name":[34],"type":["dbl"],"align":["right"]},{"label":["Value"],"name":[35],"type":["dbl"],"align":["right"]}],"data":[{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"IITA-TMS-IBA000070","12":"1929317","13":"1","14":"1","15":"112","16":"1","17":"12","18":"check","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-IITA-TMS-IBA000070_112","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"10","30":"16","31":"10","32":"10","33":"1.6","34":"1.0","35":"30.7"},{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"IITA-TMS-IBA30572","12":"1929280","13":"1","14":"1","15":"119","16":"1","17":"10","18":"check","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-IITA-TMS-IBA30572_119","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"10","30":"16","31":"10","32":"10","33":"1.6","34":"1.0","35":"32.6"},{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"IITA-TMS-IBA980581","12":"1929212","13":"1","14":"1","15":"128","16":"1","17":"1","18":"check","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-IITA-TMS-IBA980581_128","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"10","30":"16","31":"10","32":"10","33":"1.6","34":"1.0","35":"31.8"},{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"IITA-TMS-IBA982101","12":"1929309","13":"1","14":"1","15":"138","16":"1","17":"10","18":"check","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-IITA-TMS-IBA982101_138","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"8","30":"16","31":"10","32":"10","33":"1.6","34":"0.8","35":"26.8"},{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"TMEB419","12":"1929213","13":"1","14":"1","15":"116","16":"1","17":"13","18":"check","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-TMEB419_116","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"10","30":"16","31":"10","32":"10","33":"1.6","34":"1.0","35":"32.1"},{"1":"2019","2":"IITA","3":"Ubiaja","4":"19.GS.C1.C2.C3.AYT.42.UB","5":"Alpha","6":"4","7":"4","8":"0.202","9":"2019-July-09","10":"2020-July-20","11":"TMS13F1109P0009","12":"1929227","13":"1","14":"1","15":"140","16":"1","17":"12","18":"test","19":"NA","20":"NA","21":"NA","22":"2020-19.GS.C1.C2.C3.AYT.42.UB-rep1-TMS13F1109P0009_140","23":"TRUE","24":"FALSE","25":"IITA_Ubiaja_2019","26":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB","27":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1","28":"IITA_Ubiaja_2019_19.GS.C1.C2.C3.AYT.42.UB_1_1","29":"10","30":"16","31":"10","32":"10","33":"1.6","34":"1.0","35":"28.2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="considerations-before-modeling" class="section level2" number="9.4">
<h2>
<span class="header-section-number">9.4</span> Considerations before modeling<a class="anchor" aria-label="anchor" href="#considerations-before-modeling"><i class="fas fa-link"></i></a>
</h2>
<p>There is a “standard” model I have deployed reliably on Cassava MET data for the GS pipeline.</p>
<p>For the sake of demonstration, I want to pick out one trait, and do some model comparisons to illustrate the basics of selecting the best-fitting model.</p>
<p>Before doing that, a few considerations.</p>
<div id="mixed-model-software" class="section level3" number="9.4.1">
<h3>
<span class="header-section-number">9.4.1</span> Mixed-model software<a class="anchor" aria-label="anchor" href="#mixed-model-software"><i class="fas fa-link"></i></a>
</h3>
<p>In the past, I have used both <code>lmer()</code> (<code>lme4</code> R package) and <code>asreml()</code> (the R package interface for the <code>asreml</code> software) for this sort of analysis. <code>lmer()</code> is limited to homogeneous error variance structures and only certain design models / variance structures. <code>asreml()</code> is fast, flexible and industry standard <em>BUT</em> proprietary.</p>
<p>The <code>sommer</code> package is fully open-source. The <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html">sommer::mmer()</a></code> function is designed to fit almost any model <code>asreml</code> can, using a very similar syntax… but it is much more memory intensive and relatively slow. <code>sommer</code> uses direct-inversion of matrices, which have to be held in-memory in R.</p>
<p>Below, I’ll use <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html">sommer::mmer()</a></code>.</p>
<p><em>Later, we will address ways to speed up such analyses, e.g. by linking R to a multi-threaded matrix algebra library like <code>OpenBLAS</code>.</em></p>
<div id="learning-about-mixed-models" class="section level4" number="9.4.1.1">
<h4>
<span class="header-section-number">9.4.1.1</span> Learning about mixed-models<a class="anchor" aria-label="anchor" href="#learning-about-mixed-models"><i class="fas fa-link"></i></a>
</h4>
<p><strong>PLACEHOLDER FOR LINKS, REFERENCES, ETC.</strong></p>
</div>
</div>
<div id="sources-of-variability" class="section level3" number="9.4.2">
<h3>
<span class="header-section-number">9.4.2</span> Sources of variability<a class="anchor" aria-label="anchor" href="#sources-of-variability"><i class="fas fa-link"></i></a>
</h3>
<p><strong>Fixed Factors:</strong></p>
<ul>
<li>
<strong><code>yearInLoc</code></strong>: fixed-effect for year-location combination</li>
</ul>
<p><strong>Random Factors:</strong></p>
<ul>
<li>
<strong><code>germplasmName</code></strong>: random genetic effect, primary aim is to compute BLUPs for this term</li>
<li>
<strong><code>repInTrial</code></strong>: complete block replicate effects with levels uniquely nested within trials (<strong><code>studyName</code></strong><code>'s</code>)</li>
<li>
<strong><code>blockInRep</code></strong>: incomplete blocks, with levels unique nested within replicates (within <strong><code>repInTrial</code></strong><code>'s</code>)</li>
</ul>
<p><strong>Error Variance:</strong></p>
<ul>
<li>
<strong><code>studyName</code></strong>: heterogeneous error variances for each trial?</li>
<li>Another option might be to use the <strong><code>plotArea</code></strong>, or the <strong><code>plantsPerPlot</code></strong>, or the <strong><code>trialType</code></strong>
</li>
</ul>
<p><strong>“Bonus” points:</strong> If I had extra time, <strong><code>plantingDate</code></strong> and <strong><code>harvestDate</code></strong> information may provide important covariates: for example to correct yield and dry matter for trial-to-trial differences in developmental stage (age) at harvest…</p>
</div>
<div id="parallel-process-with-future" class="section level3" number="9.4.3">
<h3>
<span class="header-section-number">9.4.3</span> Parallel process with <code>future</code><a class="anchor" aria-label="anchor" href="#parallel-process-with-future"><i class="fas fa-link"></i></a>
</h3>
<p>Below, I wanted to fit 4 different models to the data. After initial tests, I felt inpatient.</p>
<p>I decided to demonstrate a version of parallel processing that most laptops should be able to do. I used the package <code>future</code> to run each model at the same time.</p>
<p>See <a href="https://cran.r-project.org/web/packages/future/vignettes/future-1-overview.html">this quick overview of <code>library(future)</code></a> to learn more.</p>
<p>Below, I choose <code>plan(multisession)</code> (activates parallel future processing), as opposed to the default, <code>plan(sequential)</code>.</p>
<p>To get the models to all run in the background, at the same time, I substitute <code>%&lt;-%{ }</code> for <code>&lt;-</code> in the code below.</p>
</div>
</div>
<div id="comparison-of-models" class="section level2" number="9.5">
<h2>
<span class="header-section-number">9.5</span> Comparison of models<a class="anchor" aria-label="anchor" href="#comparison-of-models"><i class="fas fa-link"></i></a>
</h2>
<p>What follows is a demonstration, not a complete evaluation of the options, but it shows a strategy to compare models.</p>
<p>Pick out the first chunk of <strong><code>METdata</code></strong>.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># grab just one chunk of the data to "experiment" on</span>
<span class="va">METdata</span><span class="op">&lt;-</span><span class="va">phenos</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># sommer doesn't recognize logical T/F variables</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>CompleteBlocks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">CompleteBlocks</span><span class="op">)</span>, 
            IncompleteBlocks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">IncompleteBlocks</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Not all trials (<strong><code>studyName</code></strong>s) have complete and/or incomplete blocking structures. For this reason, the “standard” model I’ve implemented for NextGen Cassava predictions over the last several years is:</p>
<ul>
<li>
<strong><code>yearInLoc</code></strong>: fixed factor</li>
<li>
<strong><code>repInTrial</code></strong> and <strong><code>blockInRep</code></strong> both random using the <code>at()</code> variance structure to model only at trials where <strong><code>CompleteBlocks=="TRUE"</code></strong>.</li>
<li>If the phenotype is a plot-basis yield component, e.g. <strong><code>logFYLD</code></strong>, include <strong><code>PropNOHAV</code></strong> as an additional fixed-effect covariate.</li>
</ul>
<div id="fit-models" class="section level3" number="9.5.1">
<h3>
<span class="header-section-number">9.5.1</span> Fit models<a class="anchor" aria-label="anchor" href="#fit-models"><i class="fas fa-link"></i></a>
</h3>
<p>The following code block implements 4 different models, each will execute in the background and at the same time.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org">future</a></span><span class="op">)</span>;
<span class="fu"><a href="https://future.futureverse.org/reference/plan.html">plan</a></span><span class="op">(</span><span class="va">multisession</span><span class="op">)</span>

<span class="co"># the model I've applied to NextGen predictions the last several years</span>
<span class="va">conv_model</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html">%&lt;-%</a></span><span class="op">{</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Value</span><span class="op">~</span><span class="va">yearInLoc</span>,
                 random<span class="op">=</span><span class="op">~</span><span class="va">germplasmName</span> <span class="op">+</span> 
                      <span class="fu">vs</span><span class="op">(</span><span class="fu">at</span><span class="op">(</span><span class="va">CompleteBlocks</span>,<span class="st">"TRUE"</span><span class="op">)</span>,<span class="va">repInTrial</span><span class="op">)</span> <span class="op">+</span> 
                      <span class="fu">vs</span><span class="op">(</span><span class="fu">at</span><span class="op">(</span><span class="va">IncompleteBlocks</span>,<span class="st">"TRUE"</span><span class="op">)</span>,<span class="va">blockInRep</span><span class="op">)</span>,
                 data<span class="op">=</span><span class="va">METdata</span>, 
                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">}</span>
<span class="co"># add heterogeneous-error variance for each trial (studyName)</span>
<span class="va">conv_het_error_model</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html">%&lt;-%</a></span> <span class="op">{</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Value</span><span class="op">~</span><span class="va">yearInLoc</span>,
                                 random<span class="op">=</span><span class="op">~</span><span class="va">germplasmName</span> <span class="op">+</span> 
                                      <span class="fu">vs</span><span class="op">(</span><span class="fu">at</span><span class="op">(</span><span class="va">CompleteBlocks</span>,<span class="st">"TRUE"</span><span class="op">)</span>,<span class="va">repInTrial</span><span class="op">)</span> <span class="op">+</span> 
                                      <span class="fu">vs</span><span class="op">(</span><span class="fu">at</span><span class="op">(</span><span class="va">IncompleteBlocks</span>,<span class="st">"TRUE"</span><span class="op">)</span>,<span class="va">blockInRep</span><span class="op">)</span>,
                                 rcov<span class="op">=</span><span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="fu">ds</span><span class="op">(</span><span class="va">studyName</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>,
                                 data<span class="op">=</span><span class="va">METdata</span>, 
                                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">}</span>
<span class="co"># simplify: no het. error, no at() variance structure</span>
<span class="va">simple_model</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html">%&lt;-%</a></span> <span class="op">{</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Value</span><span class="op">~</span><span class="va">yearInLoc</span>,
                         random<span class="op">=</span><span class="op">~</span><span class="va">germplasmName</span> <span class="op">+</span> <span class="va">repInTrial</span> <span class="op">+</span> <span class="va">blockInRep</span>,
                         data<span class="op">=</span><span class="va">METdata</span>, 
                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">}</span>
<span class="co"># no at() variance structure, include het. error variance by trial</span>
<span class="va">simple_het_error_model</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html">%&lt;-%</a></span> <span class="op">{</span> <span class="fu">mmer</span><span class="op">(</span><span class="va">Value</span><span class="op">~</span><span class="va">yearInLoc</span>,
                                   random<span class="op">=</span><span class="op">~</span><span class="va">germplasmName</span> <span class="op">+</span> <span class="va">repInTrial</span> <span class="op">+</span> <span class="va">blockInRep</span>,
                                   rcov<span class="op">=</span><span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="fu">ds</span><span class="op">(</span><span class="va">studyName</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>,
                                   data<span class="op">=</span><span class="va">METdata</span>, 
                                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">}</span></code></pre></div>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Save output of above so the above models don't need to be re-run</span>
<span class="co">## if and when I reformat or otherwise re-knit this doc to HTML</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">conv_model</span>,<span class="va">conv_het_error_model</span>,<span class="va">simple_model</span>,<span class="va">simple_het_error_model</span>,
     file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"models_compared.Rdata"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="model-comparisons" class="section level3" number="9.5.2">
<h3>
<span class="header-section-number">9.5.2</span> Model comparisons<a class="anchor" aria-label="anchor" href="#model-comparisons"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the output from the models from where I stored it</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"models_compared.Rdata"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Which model fits best? How to compare?</p>
<p><code>sommer</code> provides <code><a href="https://rdrr.io/r/stats/anova.html">anova()</a></code> for pairwise comparisons of models using a likelihood ratio test <em>and</em> by AIC, BIC.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">conv_model</span>,<span class="va">conv_het_error_model</span><span class="op">)</span>
<span class="co">#&gt; Likelihood ratio test for mixed models</span>
<span class="co">#&gt; ==============================================================</span>
<span class="co">#&gt;      Df      AIC      BIC     loLik    Chisq ChiDf</span>
<span class="co">#&gt; mod2 17 652.6920 663.2478 -324.3460               </span>
<span class="co">#&gt; mod1 10 701.9414 712.4973 -348.9707 49.24947     7</span>
<span class="co">#&gt;                      PrChisq</span>
<span class="co">#&gt; mod2                        </span>
<span class="co">#&gt; mod1 2.0273987089018e-08 ***</span>
<span class="co">#&gt; ==============================================================</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt;      Df      AIC      BIC     loLik    Chisq ChiDf</span>
<span class="co">#&gt; mod2 17 652.6920 663.2478 -324.3460               </span>
<span class="co">#&gt; mod1 10 701.9414 712.4973 -348.9707 49.24947     7</span>
<span class="co">#&gt;                      PrChisq</span>
<span class="co">#&gt; mod2                        </span>
<span class="co">#&gt; mod1 2.0273987089018e-08 ***</span></code></pre></div>
<p>Here’s a very not-pretty not-high-throughput way to set-up a table comparing the AIC for all 4 models.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>Model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"conv_model"</span>,<span class="st">"conv_het_error_model"</span>,<span class="st">"simple_model"</span>,<span class="st">"simple_het_error_model"</span><span class="op">)</span>,
       AIC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">conv_model</span><span class="op">$</span><span class="va">AIC</span>,<span class="va">conv_het_error_model</span><span class="op">$</span><span class="va">AIC</span>,<span class="va">simple_model</span><span class="op">$</span><span class="va">AIC</span>,<span class="va">simple_het_error_model</span><span class="op">$</span><span class="va">AIC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">AIC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu">gt</span><span class="fu">::</span><span class="fu"><a href="https://gt.rstudio.com/reference/gt.html">gt</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="czmmfcfmrk" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#czmmfcfmrk .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#czmmfcfmrk .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#czmmfcfmrk .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#czmmfcfmrk .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#czmmfcfmrk .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#czmmfcfmrk .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#czmmfcfmrk .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#czmmfcfmrk .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#czmmfcfmrk .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#czmmfcfmrk .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#czmmfcfmrk .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#czmmfcfmrk .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#czmmfcfmrk .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#czmmfcfmrk .gt_from_md > :first-child {
  margin-top: 0;
}

#czmmfcfmrk .gt_from_md > :last-child {
  margin-bottom: 0;
}

#czmmfcfmrk .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#czmmfcfmrk .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#czmmfcfmrk .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#czmmfcfmrk .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#czmmfcfmrk .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#czmmfcfmrk .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#czmmfcfmrk .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#czmmfcfmrk .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#czmmfcfmrk .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#czmmfcfmrk .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#czmmfcfmrk .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#czmmfcfmrk .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#czmmfcfmrk .gt_left {
  text-align: left;
}

#czmmfcfmrk .gt_center {
  text-align: center;
}

#czmmfcfmrk .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#czmmfcfmrk .gt_font_normal {
  font-weight: normal;
}

#czmmfcfmrk .gt_font_bold {
  font-weight: bold;
}

#czmmfcfmrk .gt_font_italic {
  font-style: italic;
}

#czmmfcfmrk .gt_super {
  font-size: 65%;
}

#czmmfcfmrk .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 65%;
}
</style>
<div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Model</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">AIC</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">simple_het_error_model</td>
<td class="gt_row gt_right">645.9093</td>
</tr>
<tr>
<td class="gt_row gt_left">conv_het_error_model</td>
<td class="gt_row gt_right">652.6920</td>
</tr>
<tr>
<td class="gt_row gt_left">simple_model</td>
<td class="gt_row gt_right">695.2303</td>
</tr>
<tr>
<td class="gt_row gt_left">conv_model</td>
<td class="gt_row gt_right">701.9414</td>
</tr>
</tbody>
</table></div>
</div>
<p>The “simple” heterogeneous error model fits the data best, I’ll go ahead and use it for this demonstration.</p>
<p>In addition, for plot-basis yield traits, the proportion of stands harvested (<strong>PropNOHAV</strong>) as fixed-effect covariate.</p>
</div>
</div>
<div id="learn-functions-and-iteration" class="section level2" number="9.6">
<h2>
<span class="header-section-number">9.6</span> Learn functions and iteration<a class="anchor" aria-label="anchor" href="#learn-functions-and-iteration"><i class="fas fa-link"></i></a>
</h2>
<p>At this point, taking a moment to familiarize yourself with some R packages and programming concepts might be adviseable.</p>
<ol style="list-style-type: decimal">
<li>Learn to read and write your own functions in R</li>
<li>Learn to iterate procedures in R</li>
</ol>
<div id="write-functions-in-r" class="section level3" number="9.6.1">
<h3>
<span class="header-section-number">9.6.1</span> Write functions in R<a class="anchor" aria-label="anchor" href="#write-functions-in-r"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://swcarpentry.github.io/r-novice-inflammation/02-func-R/" class="uri">https://swcarpentry.github.io/r-novice-inflammation/02-func-R/</a></p>
<p><a href="https://r4ds.had.co.nz/functions.html" class="uri">https://r4ds.had.co.nz/functions.html</a></p>
</div>
<div id="iteration-in-the-tidyverse" class="section level3" number="9.6.2">
<h3>
<span class="header-section-number">9.6.2</span> Iteration in the tidyverse<a class="anchor" aria-label="anchor" href="#iteration-in-the-tidyverse"><i class="fas fa-link"></i></a>
</h3>
<p>Packages <code>purrr</code> and (for parallel processing) <code>furrr</code>.</p>
<p><a href="https://r4ds.had.co.nz/iteration.html" class="uri">https://r4ds.had.co.nz/iteration.html</a></p>
<p><a href="https://www.r-bloggers.com/2021/09/tidy-parallel-processing-in-r-with-furrr/" class="uri">https://www.r-bloggers.com/2021/09/tidy-parallel-processing-in-r-with-furrr/</a></p>
<p><a href="https://purrr.tidyverse.org/" class="uri">https://purrr.tidyverse.org/</a></p>
<p><a href="https://dcl-prog.stanford.edu/purrr-mutate.html" class="uri">https://dcl-prog.stanford.edu/purrr-mutate.html</a></p>
<p><a href="https://dcl-prog.stanford.edu/purrr-parallel.html" class="uri">https://dcl-prog.stanford.edu/purrr-parallel.html</a></p>
<p><a href="https://furrr.futureverse.org/" class="uri">https://furrr.futureverse.org/</a></p>
</div>
</div>
<div id="analyze-all-traits" class="section level2" number="9.7">
<h2>
<span class="header-section-number">9.7</span> Analyze all traits<a class="anchor" aria-label="anchor" href="#analyze-all-traits"><i class="fas fa-link"></i></a>
</h2>
<p>Next, we will fit the “best” model from the previous step, but to all traits.</p>
<p>Below, I’ll take the time to illustrate a tidy way to set-up and then execute the model, processing traits in parallel.</p>
<p>In short, we have four data chunks set-up in the tibble <code>phenos</code></p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenos</span>
<span class="co">#&gt; # A tibble: 4 × 2</span>
<span class="co">#&gt;   Trait   METdata              </span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;list&gt;               </span>
<span class="co">#&gt; 1 DM      &lt;tibble [1,448 × 35]&gt;</span>
<span class="co">#&gt; 2 MCMDS   &lt;tibble [1,436 × 35]&gt;</span>
<span class="co">#&gt; 3 logFYLD &lt;tibble [1,590 × 35]&gt;</span>
<span class="co">#&gt; 4 logDYLD &lt;tibble [1,443 × 35]&gt;</span></code></pre></div>
<p>Let’s go step-by-step to build a good understanding of a programming workflow. We’ll use similar procedures in many of the analyses and data processes we do downstream.</p>
<ol style="list-style-type: decimal">
<li>We will write a function that will take each data chunk (tibbles stored in <code>phenos$METdata</code>) as an input argument.</li>
<li>We’ll execute the function (and thus the analysis) simultaneously (in parallel) across the data chunks using the <code>furrr</code> package.</li>
</ol>
<div id="test-code-for-function" class="section level3" number="9.7.1">
<h3>
<span class="header-section-number">9.7.1</span> Test code for function<a class="anchor" aria-label="anchor" href="#test-code-for-function"><i class="fas fa-link"></i></a>
</h3>
<p>Here’s how we want our call to <code>mmer()</code> to fit the mixed-model to look like:</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># DON"T RUN</span>
<span class="va">simple_het_error_model</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="va">Value</span><span class="op">~</span><span class="va">yearInLoc</span>,
                             random<span class="op">=</span><span class="op">~</span><span class="va">germplasmName</span> <span class="op">+</span> <span class="va">repInTrial</span> <span class="op">+</span> <span class="va">blockInRep</span>,
                             rcov<span class="op">=</span><span class="op">~</span><span class="fu">vs</span><span class="op">(</span><span class="fu">ds</span><span class="op">(</span><span class="va">studyName</span><span class="op">)</span>,<span class="va">units</span><span class="op">)</span>,
                             data<span class="op">=</span><span class="va">METdata</span>, getPEV <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>To build a function around this, we want to be able to input different values ofthe arguments to <code>mmer()</code>, namely: <code>data=</code>, <code>fixed=</code>, <code>random=</code> and <code>rcov=</code>.</p>
<p>Note that in this case, the only actual difference among traits in our example is going to be an added fixed effect for the yield traits, the <strong><code>PropNOHAV</code></strong>.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Input arguments for the function:</span>
<span class="va">Trait</span><span class="op">&lt;-</span><span class="va">phenos</span><span class="op">$</span><span class="va">Trait</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> 
<span class="co"># input for "data="</span>
<span class="va">METdata</span><span class="op">&lt;-</span><span class="va">phenos</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co"># input for "fixed="</span>
<span class="va">fixedFormula</span><span class="op">&lt;-</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Trait</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"logDYLD"</span>,<span class="st">"logFYLD"</span>,<span class="st">"logRTNO"</span>,<span class="st">"logTOPYLD"</span><span class="op">)</span>,
                     <span class="st">"Value ~ yearInLoc + PropNOHAV"</span>,
                     <span class="st">"Value ~ yearInLoc"</span><span class="op">)</span>
<span class="co"># input for "random="</span>
<span class="va">randFormula</span><span class="op">&lt;-</span><span class="st">"~germplasmName + repInTrial + blockInRep"</span>;
<span class="co"># input for "rcov="</span>
<span class="va">rcovFormula</span><span class="op">=</span><span class="st">"~vs(ds(studyName),units)"</span></code></pre></div>
<p>Here’s how we can call <code>mmer()</code> inside another function, using the variables in the above code chunk as input arguments:</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">starttime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>
<span class="va">mmer_output</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">fixedFormula</span><span class="op">)</span>,
                  random <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">randFormula</span><span class="op">)</span>,
                  rcov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">rcovFormula</span><span class="op">)</span>,
                  data<span class="op">=</span><span class="va">METdata</span>, 
                  getPEV <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="va">stoptime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>; <span class="va">elapsed</span><span class="op">&lt;-</span><span class="va">stoptime</span><span class="op">-</span><span class="va">starttime</span>; <span class="va">elapsed</span><span class="op">/</span><span class="fl">60</span>
<span class="co"># iteration    LogLik     wall    cpu(sec)   restrained</span>
<span class="co">#     1      -367.006   10:18:35      31           0</span>
<span class="co">#     2      -338.545   10:19:5      61           0</span>
<span class="co">#     3      -325.9   10:19:38      94           0</span>
<span class="co">#     4      -321.988   10:20:8      124           0</span>
<span class="co">#     5      -321.2   10:20:37      153           0</span>
<span class="co">#     6      -321.017   10:21:4      180           0</span>
<span class="co">#     7      -320.971   10:21:29      205           0</span>
<span class="co">#     8      -320.959   10:21:55      231           0</span>
<span class="co">#     9      -320.956   10:22:20      256           0</span>
<span class="co">#     10      -320.955   10:22:46      282           0</span>
<span class="co">#  elapsed </span>
<span class="co"># 4.775417 </span></code></pre></div>
<p>Took 5 whole minutes to fit that, yikes.</p>
<p><strong>Outlier Detection and Removal:</strong> My standard pipeline includes outlier detection and removal. (1) Fit the model. (2) Define outliers as observations where the absolute-value of the standardized residuals &gt;3.3. (3) Remove outliers and re-fit the model. (4) Check for and remove residuals again. (5) If more residuals were detected and removed, re-fit a final time.</p>
<p>For the sake of a simple, quick example, the demonstration below, will just do one round removing outliers.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># index observations that are defined as outliers</span>
<span class="va">outliers</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">mmer_output</span><span class="op">$</span><span class="va">residuals</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">3.3</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">outliers</span><span class="op">)</span> <span class="co"># How many?</span>
<span class="co"># [1] 6</span></code></pre></div>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remove outliers, if any</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">outliers</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="op">{</span> 
     <span class="va">x</span><span class="op">&lt;-</span><span class="va">METdata</span><span class="op">[</span><span class="op">-</span><span class="va">outliers</span>,<span class="op">]</span> 
     <span class="co"># Refit the model</span>
     <span class="va">starttime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>
     <span class="va">mmer_output</span><span class="op">&lt;-</span><span class="fu">mmer</span><span class="op">(</span>fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">fixedFormula</span><span class="op">)</span>,
                              random <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">randFormula</span><span class="op">)</span>,
                              rcov<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/formula.html">as.formula</a></span><span class="op">(</span><span class="va">rcovFormula</span><span class="op">)</span>,
                              data<span class="op">=</span><span class="va">x</span>, 
                              getPEV <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
     <span class="va">stoptime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>; <span class="va">elapsed</span><span class="op">&lt;-</span><span class="va">stoptime</span><span class="op">-</span><span class="va">starttime</span>; <span class="va">elapsed</span><span class="op">/</span><span class="fl">60</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">outliers</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">{</span> <span class="va">outliers</span><span class="op">&lt;-</span><span class="cn">NULL</span> <span class="op">}</span>

<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">mmer_output</span>,<span class="va">outliers</span>,
     file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"mmer_output_example_during_stage1.Rdata"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Next we extract and format the outputs we want from the mixed-model; these will be combined into a list and passed to the <code><a href="https://rdrr.io/r/base/function.html">return()</a></code> statement of the function we are building.</p>
<p>In the chunk below, I extract and compute a number of things:</p>
<ol style="list-style-type: decimal">
<li><p>Model fit stats like AIC, log likelihood and the number of grouping factors, mostly stored for posterity</p></li>
<li><p>Variance component estimates</p></li>
<li><p>A broad-sense heritability value <code>H2</code></p></li>
<li><p>Extract <strong><code>BLUP</code></strong>s and corresponding <strong><code>PEV</code></strong>s (prediction error variance) for the genotype-factor (“<strong>germplasmName</strong>”)</p></li>
<li>
<p>Compute</p>
<ul>
<li><p>reliabilities (<strong><code>REL</code></strong>) for each <strong><code>BLUP</code></strong></p></li>
<li><p>de-regressed BLUPs (<strong><code>drgBLUP</code></strong>)</p></li>
<li><p>Weighting factor (<strong><code>WT</code></strong>) for each BLUP, to be used in downstream analyses where the <strong><code>drgBLUP</code></strong> is supplied as the response variable.</p></li>
</ul>
</li>
</ol>
<p>In a few sections (<a href="preliminary-field-trial-analysis.html#de-regression-explained">see here</a>), I’ll break down what these are and how they are used.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"mmer_output_example_during_stage1.Rdata"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># summary(mmer_output)</span>

<span class="co"># log likelihood of the model, AIC, convergence T/F</span>
<span class="va">modelfit</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mmer_output</span><span class="op">)</span><span class="op">$</span><span class="va">logo</span>
<span class="co"># number of groups for factors, could be used to compute DF </span>
<span class="va">groups</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mmer_output</span><span class="op">)</span><span class="op">$</span><span class="va">groups</span>
<span class="co"># variance components</span>
<span class="va">varcomp</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mmer_output</span><span class="op">)</span><span class="op">$</span><span class="va">varcomp</span>

<span class="co"># genotypic variance</span>
<span class="va">Vg</span><span class="op">&lt;-</span><span class="va">mmer_output</span><span class="op">$</span><span class="va">sigma</span><span class="op">$</span><span class="va">germplasmName</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># Mean error-variance across trials </span>
<span class="co">## across heterogeneous error variance estimates</span>
<span class="va">meanHetVarE</span><span class="op">&lt;-</span><span class="va">mmer_output</span><span class="op">$</span><span class="va">sigma</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">":units"</span>,<span class="va">.</span>,value<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># Mean number of reps per accession</span>
<span class="va">meanNreps</span><span class="op">&lt;-</span><span class="va">METdata</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">germplasmName</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/exposition.html">%$%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>
<span class="co"># Broad-sense heritability</span>
<span class="va">H2</span><span class="op">&lt;-</span><span class="va">Vg</span><span class="op">/</span><span class="op">(</span><span class="va">Vg</span><span class="op">+</span><span class="op">(</span><span class="va">meanHetVarE</span><span class="op">/</span><span class="va">meanNreps</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Extract the BLUPs and PEVs, compute Reliability (REL), </span>
<span class="co"># de-regressed BLUPs and weights for downstream analysis</span>
<span class="va">blups</span><span class="op">&lt;-</span><span class="va">mmer_output</span><span class="op">$</span><span class="va">U</span><span class="op">$</span><span class="va">germplasmName</span><span class="op">$</span><span class="va">Value</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>germplasmName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>,BLUP<span class="op">=</span><span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>germplasmName<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"germplasmName"</span>,<span class="st">""</span>,<span class="va">germplasmName</span><span class="op">)</span>,
            <span class="co"># prediction error variance</span>
            PEV<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">mmer_output</span><span class="op">$</span><span class="va">PevU</span><span class="op">$</span><span class="va">germplasmName</span><span class="op">$</span><span class="va">Value</span><span class="op">)</span>, 
            REL<span class="op">=</span><span class="fl">1</span><span class="op">-</span><span class="va">PEV</span><span class="op">/</span><span class="va">Vg</span>, <span class="co"># Reliability</span>
            drgBLUP<span class="op">=</span><span class="va">BLUP</span><span class="op">/</span><span class="va">REL</span>, <span class="co"># De-regressed BLUP</span>
            WT<span class="op">=</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">H2</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="op">(</span><span class="fl">0.1</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">REL</span><span class="op">)</span><span class="op">/</span><span class="va">REL</span><span class="op">)</span><span class="op">*</span><span class="va">H2</span><span class="op">)</span> <span class="co"># weight for downstream</span>
            <span class="op">)</span>

<span class="co"># Combine all outputs into one object the function can return() </span>
<span class="va">out</span><span class="op">&lt;-</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>H2<span class="op">=</span><span class="va">H2</span>,meanHetVarE<span class="op">=</span><span class="va">meanHetVarE</span>,meanNreps<span class="op">=</span><span class="va">meanNreps</span>,
            modelfit<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">modelfit</span><span class="op">)</span>,
            groups<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">groups</span><span class="op">)</span>,
            blups<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">blups</span><span class="op">)</span>,
            varcomp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">varcomp</span><span class="op">)</span>,
            outliers<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">outliers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="set-up-the-function" class="section level3" number="9.7.2">
<h3>
<span class="header-section-number">9.7.2</span> Set-up the function<a class="anchor" aria-label="anchor" href="#set-up-the-function"><i class="fas fa-link"></i></a>
</h3>
<p>I combined the code chunks above into a function <strong><code>get_blups()</code></strong>. To spare the duplication of code here, the function is in its own R script: <strong><code>code/get_blups.R</code></strong>.</p>
<p><strong>A NOTE ON DEBUGGING:</strong> Building and executing analyses in this way is often an iterative process. After my first run of the function <code>get_blups()</code>, I noticed all the results were the same. In this case, I leftout an argument from the <code>function()</code> call. Details not important. <em>SUGGESTION</em>: When you put the code into an actual <code>function()</code>, test it on a single code chunk, especially if the analysis will run a long time or use large resources.</p>
</div>
<div id="run-the-full-analysis" class="section level3" number="9.7.3">
<h3>
<span class="header-section-number">9.7.3</span> Run the full analysis<a class="anchor" aria-label="anchor" href="#run-the-full-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>Finally, we are ready to source the function we built and execute the full analysis on all traits with it.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"code"</span>,<span class="st">"get_blups.R"</span><span class="op">)</span><span class="op">)</span>
<span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/compound.html">%&lt;&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>fixedFormula<span class="op">=</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fifelse.html">ifelse</a></span><span class="op">(</span><span class="va">Trait</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"logDYLD"</span>,<span class="st">"logFYLD"</span>,<span class="st">"logRTNO"</span>,<span class="st">"logTOPYLD"</span><span class="op">)</span>,
                                <span class="st">"Value ~ yearInLoc + PropNOHAV"</span>,
                                <span class="st">"Value ~ yearInLoc"</span><span class="op">)</span>,
            randFormula<span class="op">=</span><span class="st">"~germplasmName + repInTrial + blockInRep"</span>,
            rcovFormula<span class="op">=</span><span class="st">"~vs(ds(studyName),units)"</span><span class="op">)</span></code></pre></div>
<p>The next chunk uses the function <code><a href="https://rdrr.io/pkg/furrr/man/future_map2.html">future_pmap()</a></code> from the <code>furrr</code> package to exectute the <code>get_blups()</code> function across all four traits in parallel.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">starttime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisVaughan/furrr">furrr</a></span><span class="op">)</span>; <span class="fu"><a href="https://future.futureverse.org/reference/plan.html">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">blups</span><span class="op">&lt;-</span><span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>modelOut<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/furrr/man/future_map2.html">future_pmap</a></span><span class="op">(</span><span class="va">.</span>,<span class="va">get_blups</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">modelOut</span><span class="op">)</span>
<span class="va">stoptime</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>; <span class="va">elapsed</span><span class="op">&lt;-</span><span class="va">stoptime</span><span class="op">-</span><span class="va">starttime</span>; <span class="va">elapsed</span><span class="op">/</span><span class="fl">60</span>
<span class="co">#  elapsed </span>
<span class="co"># 17.25957 </span>
<span class="co"># Save</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">blups</span>,file <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Took 17 minutes to run on my laptop :(</p>
</div>
<div id="why-is-mcmds-heritability-0" class="section level3" number="9.7.4">
<h3>
<span class="header-section-number">9.7.4</span> Why is MCMDS heritability 0?<a class="anchor" aria-label="anchor" href="#why-is-mcmds-heritability-0"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># blups %&gt;% rmarkdown::paged_table()</span>
<span class="va">blups</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">H2</span>,<span class="va">meanHetVarE</span>,<span class="va">meanNreps</span>,<span class="va">outliers</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>H2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">H2</span>,<span class="fl">3</span><span class="op">)</span>,
            meanNreps<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">meanNreps</span>,<span class="fl">3</span><span class="op">)</span>,
            outliers<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dbl</a></span><span class="op">(</span><span class="va">outliers</span>,<span class="va">length</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Trait"],"name":[1],"type":["chr"],"align":["left"]},{"label":["H2"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["meanHetVarE"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["meanNreps"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["outliers"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"DM","2":"0.578","3":"6.0863897","4":"1.785","5":"6"},{"1":"MCMDS","2":"0.000","3":"0.2048047","4":"1.520","5":"33"},{"1":"logFYLD","2":"0.300","3":"0.1289105","4":"1.719","5":"9"},{"1":"logDYLD","2":"0.347","3":"0.1119542","4":"1.784","5":"4"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Why is the H2 for <strong>MCMDS</strong> 0? Aside from this being an unfortunate result, knowing that it is unusual / atypical requires knowledge of cassava genetic architecture.</p>
<p>The heritablity of <strong>DM</strong>, <strong>logFYLD</strong> and <strong>logDYLD</strong> are both pretty much in line with “typical” results, <em>but</em> MCMDS in most cassava populations is a pretty heritable trait.</p>
<p>I notice there were 33 outliers for MCMDS.</p>
<p>Let’s look at the distribution of MCMDS in the raw data:</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-16-1.png" width="672"></div>
<p>This is actually often how MCMDS scores look like.</p>
<p>Which are the observations that were removed as outliers?</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">blups</span><span class="op">$</span><span class="va">outliers</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/exposition.html">%$%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">Value</span><span class="op">)</span>
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;    3.50    4.00    4.00    4.03    4.00    5.00</span></code></pre></div>
<p>They are exclusively high scores.</p>
<p>I wonder if removing them was a mistake:</p>
<p>To find out, let’s re-run the analysis just for MCMDS, this time turning outlier removal off. I actually added an <code>removeOutliers=T/F</code> argument and a few lines of code to the <code>get_blups()</code> function to make this happen:</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"code"</span>,<span class="st">"get_blups.R"</span><span class="op">)</span><span class="op">)</span>
<span class="va">blups_mcmds_noRemoveOut</span><span class="op">&lt;-</span><span class="va">phenos</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Trait</span><span class="op">==</span><span class="st">"MCMDS"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>modelOut<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">pmap</a></span><span class="op">(</span><span class="va">.</span>,<span class="va">get_blups</span>,removeOutliers <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">modelOut</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">blups_mcmds_noRemoveOut</span>,
     file<span class="op">=</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups_mcmds_noRemoveOut.Rdata"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"output"</span>,<span class="st">"blups_mcmds_noRemoveOut.Rdata"</span><span class="op">)</span><span class="op">)</span>
<span class="va">blups_mcmds_noRemoveOut</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">Trait</span>,<span class="va">H2</span>,<span class="va">meanHetVarE</span>,<span class="va">meanNreps</span>,<span class="va">outliers</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>H2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">H2</span>,<span class="fl">3</span><span class="op">)</span>,
            meanNreps<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">meanNreps</span>,<span class="fl">3</span><span class="op">)</span>,
            outliers<span class="op">=</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dbl</a></span><span class="op">(</span><span class="va">outliers</span>,<span class="va">length</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Trait"],"name":[1],"type":["chr"],"align":["left"]},{"label":["H2"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["meanHetVarE"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["meanNreps"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["outliers"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"MCMDS","2":"0","3":"0.7744563","4":"1.52","5":"0"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups_mcmds_noRemoveOut</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;  germplasmName           BLUP           </span>
<span class="co">#&gt;  Length:945         Min.   :-3.112e-26  </span>
<span class="co">#&gt;  Class :character   1st Qu.:-1.212e-27  </span>
<span class="co">#&gt;  Mode  :character   Median :-8.745e-28  </span>
<span class="co">#&gt;                     Mean   : 2.747e-29  </span>
<span class="co">#&gt;                     3rd Qu.:-3.398e-28  </span>
<span class="co">#&gt;                     Max.   : 1.143e-25  </span>
<span class="co">#&gt;       PEV                 REL           </span>
<span class="co">#&gt;  Min.   :2.301e-27   Min.   :0.000e+00  </span>
<span class="co">#&gt;  1st Qu.:2.301e-27   1st Qu.:0.000e+00  </span>
<span class="co">#&gt;  Median :2.301e-27   Median :0.000e+00  </span>
<span class="co">#&gt;  Mean   :2.301e-27   Mean   :1.096e-11  </span>
<span class="co">#&gt;  3rd Qu.:2.301e-27   3rd Qu.:0.000e+00  </span>
<span class="co">#&gt;  Max.   :2.301e-27   Max.   :4.742e-10  </span>
<span class="co">#&gt;     drgBLUP                 WT           </span>
<span class="co">#&gt;  Min.   :      -Inf   Min.   :0.000e+00  </span>
<span class="co">#&gt;  1st Qu.:      -Inf   1st Qu.:0.000e+00  </span>
<span class="co">#&gt;  Median :      -Inf   Median :0.000e+00  </span>
<span class="co">#&gt;  Mean   :       NaN   Mean   :2.426e+15  </span>
<span class="co">#&gt;  3rd Qu.:-6.563e-17   3rd Qu.:0.000e+00  </span>
<span class="co">#&gt;  Max.   :       Inf   Max.   :1.050e+17</span></code></pre></div>
<p><strong>DECLARING MCMDS A LOSS</strong> You can see from the summary, MCMDS <em>in this dataset</em> have essentially no genetic variance / reliability. For the sake of moving on with the example, I will exclude MCMDS from further consideration.</p>
<p>For my breeding program or critical experiment, I <em>would not</em> stop there. If there are not alternative trials to draw data from, the non-normally distributed, pseudo-count distributed MCMDS data can be modeled as such.</p>
</div>
</div>
<div id="de-regression-explained" class="section level2" number="9.8">
<h2>
<span class="header-section-number">9.8</span> De-regressed BLUPs and weighted two-stage prediction<a class="anchor" aria-label="anchor" href="#de-regression-explained"><i class="fas fa-link"></i></a>
</h2>
<p>The <strong><code>BLUP</code></strong> represents a single performance prediction for each unique <strong>germplasmName</strong> <em>without</em> accounting for genomic relatedness using marker data but accounting for the potentially unbalanced original trial data’s experimental design factors. The <strong>BLUP</strong>s are in the units of the original trait but centered to the population mean.</p>
<p><strong>BLUP</strong>s are shrunken, <em>i.e.</em> regressed towards the mean.Downstream genomic predictions fit the <strong>germplasmName</strong> again as a random effect in a mixed-model. For this reasion, using <strong>BLUP</strong>s directly as the response variable would doubly shrink the data. Instead, it is advised to do what is called “de-regression” of the BLUPs <span class="citation">Garrick <em>et al.</em> (2009)</span>.</p>
<p>The de-regressed BLUP (<strong>drgBLUP</strong>) is simply the BLUP divided by a quantity called the <strong>reliability</strong> or <span class="math inline">\(r^{2}\)</span>.</p>
<p>The reliability is defined as: <span class="math inline">\(r^{2} = 1 - \frac{PEV}{\sigma^{2}_{g}}\)</span></p>
<ul>
<li>
<strong>PEV:</strong> = prediction error variance, a measure exactly what it sounds like, the error variance associated with each individual prediction, which may be different among predictions, depending on the data. When we ran <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html">sommer::mmer()</a></code> we set <code>getPEV=TRUE</code> and then later extracted the diagonal-values of the resulting PEV matrix (<code>PEV=diag(mmer_output$PevU$germplasmName$Value)</code>).</li>
<li>
<span class="math inline">\(\sigma^{2}_{g}\)</span> = the genetic variance estimated in Step 1. <strong>PEV</strong> will always be some fraction of the total genetic variance.</li>
</ul>
<p>Thus reliability ranges from 0 to 1 and measures the <em>certainty</em> surrounding each <strong><code>BLUP</code></strong> value; or rather, the probability the <strong><code>BLUP</code></strong> would change if anothe experiment (data point) were added. So if <span class="math inline">\(r^{2}=1\)</span>, then we have essentially zero expected error for a BLUP. For example, check varieties often have so many observations in a dataset that they will have high reliability.</p>
<p>So if a de-regressed BLUP is defined as <span class="math inline">\(\frac{BLUP}{r^{2}}\)</span> you can see that for clones with high reliability, the BLUP will stay essentially the same, but if the reliability is low, the <strong><code>BLUP</code></strong> will be <em>un-shrink</em> or <em>de-regress</em> strongly. This is going to happen because genotypes with few observations and thus low reliability will be more strongly shrunken to zero during the first step.</p>
<p>Below is a density plot showing the distribution of the original data compared to the <strong><code>BLUP</code></strong> and the <strong><code>drgBLUP</code></strong>. This is meant to show how shrinkage and unshrinkage effect the data. Since BLUPs and drgBLUPs are centered on the mean, I added back the grand mean for comparison to the original data.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">popMean</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Value</span><span class="op">)</span>
<span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">observationUnitDbId</span>,<span class="va">Value</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Data<span class="op">=</span><span class="st">"Original Data"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">germplasmName</span>,<span class="va">BLUP</span>,<span class="va">drgBLUP</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
                    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BLUP"</span>,<span class="st">"drgBLUP"</span><span class="op">)</span>,
                                 names_to <span class="op">=</span> <span class="st">"Data"</span>, values_to <span class="op">=</span> <span class="st">"Value"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Value<span class="op">=</span><span class="va">Value</span><span class="op">+</span><span class="va">popMean</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Value</span>,fill<span class="op">=</span><span class="va">Data</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">popMean</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-21-1.png" width="672"></div>
<p>Next, I grab the blups from only the <strong>germplasmName</strong> with the two highest and the two lowest <strong>REL</strong>. In all cases, the de-regressed BLUP should shift away from the mean, but proportional to the reliability. Below, I devised a plot to show that:</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>
<span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">REL</span>,n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_min</a></span><span class="op">(</span>order_by <span class="op">=</span> <span class="va">REL</span>,n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">germplasmName</span>,<span class="va">REL</span>,<span class="va">BLUP</span>,<span class="va">drgBLUP</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BLUP"</span>,<span class="st">"drgBLUP"</span><span class="op">)</span>,
                  names_to <span class="op">=</span> <span class="st">"Data"</span>,values_to <span class="op">=</span> <span class="st">"Value"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Value</span>,y<span class="op">=</span><span class="va">germplasmName</span>,
                  color<span class="op">=</span><span class="va">REL</span>,group<span class="op">=</span><span class="va">germplasmName</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>shape<span class="op">=</span><span class="va">Data</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label<span class="op">=</span><span class="va">Data</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> </code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-22-1.png" width="672"></div>
<p>The de-regressed BLUP will be used as a response variable in downstream genomic prediction.</p>
<p>We can account for the differential reliability among genotypes downstream using a weighted analysis. We will compute our weights according to <span class="citation">Garrick <em>et al.</em> (2009)</span>. Using <code><a href="https://rdrr.io/pkg/sommer/man/mmer.html">sommer::mmer()</a></code>, weights can be supplied through the <code>weights=</code> argument (see the <a href="https://cran.r-project.org/web/packages/sommer/sommer.pdf">sommer::mmer documentation</a> for details). The values supplied will be weight the error term proportional to <span class="math inline">\(\frac{1}{\sqrt{WT}}\)</span>, meaning drgBLUP with higher WT have smaller error; more effect on the prediction.</p>
<p><span class="citation">Garrick <em>et al.</em> (2009)</span>’s recommended weight is:</p>
<p><span class="math display">\[WT = \frac{1-H^{2}}{0.1+\frac{1-r^{2}}{r^{r}} \times H^{2}}\]</span></p>
<ul>
<li>
<span class="math inline">\(H^{2}\)</span> is the broad-sense heritability (H2).</li>
<li>
<span class="math inline">\(r^{2}\)</span> is reliability (REL).</li>
</ul>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">REL</span><span class="op">&lt;-</span><span class="va">H2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from<span class="op">=</span><span class="fl">0.1</span>,to<span class="op">=</span><span class="fl">0.9</span>,by<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span>REL<span class="op">=</span><span class="va">REL</span>,H2<span class="op">=</span><span class="va">H2</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>WT<span class="op">=</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">H2</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="op">(</span><span class="fl">0.1</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">REL</span><span class="op">)</span><span class="op">/</span><span class="va">REL</span><span class="op">)</span><span class="op">*</span><span class="va">H2</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">REL</span>,y<span class="op">=</span><span class="va">WT</span>,color<span class="op">=</span><span class="va">H2</span>, group<span class="op">=</span><span class="va">H2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="op">)</span> </code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-23-1.png" width="672"></div>
<p>I made a plot showing how <strong><code>WT</code></strong> varies across a range of <strong><code>H2</code></strong> and <strong><code>REL</code></strong>. This is to show how the weighting behaves. As we see, <strong><code>WT</code></strong> scales with <strong><code>REL</code></strong> and <strong><code>H2</code></strong>. <strong><code>H2</code></strong> conditions the overall variability among individuals in their weight in the analysis. This works such that if heritability is low, there is a greater relative premium (weight) placed on individuals with high quality data (high <strong><code>REL</code></strong>), usually those with high numbers of observations, e.g. check-genotypes. If heritability is high, <strong><code>WT</code></strong>s are low and there isn’t much difference between low and high <strong><code>REL</code></strong>.</p>
<p>Here’s a plot that shows from the actual <code>blups</code>, how the <strong><code>WT</code></strong> scales with the number of observations-per-genotype.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># Count the number of plots per clone in the raw data and merge to the BlUPs</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">germplasmName</span>,name <span class="op">=</span> <span class="st">"Nplots"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Nplots</span>,y<span class="op">=</span><span class="va">WT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "germplasmName"</span></code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-24-1.png" width="672"></div>
<p>Next, I plot the <strong><code>BLUP</code></strong>s again the de-regressed BLUPs and scale point-size (<strong><code>WT</code></strong>) and color (<strong><code>REL</code></strong>). The red line has a slope of 1. You can see clearly how high reliability predictions don’t change much (close to the red line). You can also see that the farther the <strong><code>BLUP</code></strong> from the mean, the more the de-regression (the more the original shrinkage).</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">blups</span><span class="op">$</span><span class="va">blups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="co"># Count the number of plots per clone in the raw data and merge to the BlUPs</span>
     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">blups</span><span class="op">$</span><span class="va">METdata</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">germplasmName</span>,name <span class="op">=</span> <span class="st">"Nplots"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/genomicMateSelectR/man/pipe.html">%&gt;%</a></span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">BLUP</span>,y<span class="op">=</span><span class="va">drgBLUP</span>,color<span class="op">=</span><span class="va">REL</span>,size<span class="op">=</span><span class="va">WT</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_abline</a></span><span class="op">(</span>slope<span class="op">=</span><span class="fl">1</span>, color<span class="op">=</span><span class="st">'darkred'</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = "germplasmName"</span></code></pre></div>
<div class="inline-figure"><img src="05-getBLUPs_files/figure-html/unnamed-chunk-25-1.png" width="672"></div>
<div id="computing-h2-for-weights" class="section level3" number="9.8.1">
<h3>
<span class="header-section-number">9.8.1</span> Computing H2 for weights<a class="anchor" aria-label="anchor" href="#computing-h2-for-weights"><i class="fas fa-link"></i></a>
</h3>
<p>Quick note on computing <span class="math inline">\(H^2\)</span> for the <strong><code>WT</code></strong>s above. Heterogeneneous error variance model does not lend itself to standard compute of <span class="math inline">\(H^2\)</span> across trials.</p>
<p>We need an <span class="math inline">\(H^2\)</span> that reflects the overall dataset. The analysis above used heterogenoeous error variances. Therefore, I decided to use the following <span class="math inline">\(H^2\)</span>:</p>
<p><span class="math display">\[H^2 = \frac{\sigma^2_g}{\sigma^2_g+\bar{\sigma^2_{e}}/\bar{n}_{reps}}\]</span></p>
<p>The <span class="math inline">\(\sigma^2_g\)</span> is the mixed-model estimate.</p>
<p>The <span class="math inline">\(\bar{\sigma^2_{e}}\)</span> is the average error-variance estimate across trials.</p>
<p><span class="math inline">\(\bar{n}_{reps}\)</span> is the mean number of reps per genotype in the dataset.</p>
<p>So this <span class="math inline">\(H^2\)</span> definition should reflect the average trial and the average number of replicates per line.</p>
<p>Alternatives are definitely possible and users are encouraged to question and experiment!</p>
</div>
</div>
<div id="mixed-models-in-brief" class="section level2" number="9.9">
<h2>
<span class="header-section-number">9.9</span> Mixed models in brief<a class="anchor" aria-label="anchor" href="#mixed-models-in-brief"><i class="fas fa-link"></i></a>
</h2>
<p>There are often factors that effect our experiments that we don’t specifically care about. As an example, we might replicate our experiment within a site by planting two different fields with the same design. We can easily imagine that small variations in soil, slope and aspect, elevation and planting date might influence the outcome at least to some extent. But most of the time, we aren’t after the effect of the field or planting date. We want to learn some general about the location as a whole.</p>
<p>We can make some assumption that there are a universe of possible outcomes from any given part of a field station, for example, that might be observed. Every replicate / experiment we do just shows us one manifestation of that.</p>
<p>This makes the replication, field or planting date a <strong>random variable</strong>.</p>
<p>In addition, data are often complex and messy. Sample sizes in individual grouping variables often leave something to be desired , especially if we are trying to fit complicated models with many parameters. On top of that our data points might not be truly independent - for instance, there could be spatial autocorrelation between plots or blocks that are more similar because they are in a similar location. Or, as is the case in genomic selection, there we might have information about the genetic relatedness of the germplasm we are growing, that we can measure and want to account for.</p>
<p>This is why mixed models were developed, to deal with such messy data and to allow us to use all our data, even when we have low sample sizes, structured data, and many covariates to fit. Oh, and on top of all that mixed models allow us to save degrees of freedom compared to running standard linear models! Sounds good, doesn’t it?</p>
<p>In broad terms, <strong>fixed effects</strong> are variables that we expect will have an effect on the dependent/response variable. Most of the time, <strong>random effects</strong> are usually grouping factors for which we are trying to control. A lot of the time we are not specifically interested in their impact on the response variable. Additionally, the data for our random effect is just a sample of all the possibilities.</p>
<p>The above three paragraphs were paraphrased from (<a href="https://ourcodingclub.github.io/2017/03/15/mixed-models.html" class="uri">https://ourcodingclub.github.io/2017/03/15/mixed-models.html</a>).</p>
<p>There is a <strong>major</strong> exception to the statement above about using random-effects to control e.g. replication variance: genotypes.</p>
<p>Our primary objective is to obtain correct rankings of the genetic- or heritable-economic value of each accession, in order to determine which to cross or advance.</p>
<div id="links-for-learning" class="section level3" number="9.9.1">
<h3>
<span class="header-section-number">9.9.1</span> Links for learning<a class="anchor" aria-label="anchor" href="#links-for-learning"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://si.biostat.washington.edu/sites/default/files/modules/Seattle-SISG-18-IntroQG-Lecture08.pdf" class="uri">https://si.biostat.washington.edu/sites/default/files/modules/Seattle-SISG-18-IntroQG-Lecture08.pdf</a></p>
<p><a href="https://towardsdatascience.com/how-linear-mixed-model-works-350950a82911" class="uri">https://towardsdatascience.com/how-linear-mixed-model-works-350950a82911</a></p>
<p><a href="https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/" class="uri">https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/</a></p>
<p><a href="https://www.youtube.com/playlist?list=PLZ0lafzH_UmfHsoePj0t7xQglKSOpzbDp">Excellence in Breeding - Basics of GS - 2021 YouTube Video Series with Eduardo Covarrubias</a></p>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="prepare-genotypic-data.html"><span class="header-section-number">8</span> Prepare genotypic data</a></div>
<div class="next"><a href="intro-to-genomic-prediction.html"><span class="header-section-number">10</span> Intro to Genomic Prediction</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#preliminary-field-trial-analysis"><span class="header-section-number">9</span> Preliminary field trial analysis</a></li>
<li><a class="nav-link" href="#process-map-3"><span class="header-section-number">9.1</span> Process Map</a></li>
<li><a class="nav-link" href="#one-stage-or-multi-stage"><span class="header-section-number">9.2</span> One-stage or multi-stage?</a></li>
<li><a class="nav-link" href="#set-up-training-datasets"><span class="header-section-number">9.3</span> Set-up training datasets</a></li>
<li>
<a class="nav-link" href="#considerations-before-modeling"><span class="header-section-number">9.4</span> Considerations before modeling</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#mixed-model-software"><span class="header-section-number">9.4.1</span> Mixed-model software</a></li>
<li><a class="nav-link" href="#sources-of-variability"><span class="header-section-number">9.4.2</span> Sources of variability</a></li>
<li><a class="nav-link" href="#parallel-process-with-future"><span class="header-section-number">9.4.3</span> Parallel process with future</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#comparison-of-models"><span class="header-section-number">9.5</span> Comparison of models</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#fit-models"><span class="header-section-number">9.5.1</span> Fit models</a></li>
<li><a class="nav-link" href="#model-comparisons"><span class="header-section-number">9.5.2</span> Model comparisons</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#learn-functions-and-iteration"><span class="header-section-number">9.6</span> Learn functions and iteration</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#write-functions-in-r"><span class="header-section-number">9.6.1</span> Write functions in R</a></li>
<li><a class="nav-link" href="#iteration-in-the-tidyverse"><span class="header-section-number">9.6.2</span> Iteration in the tidyverse</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#analyze-all-traits"><span class="header-section-number">9.7</span> Analyze all traits</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#test-code-for-function"><span class="header-section-number">9.7.1</span> Test code for function</a></li>
<li><a class="nav-link" href="#set-up-the-function"><span class="header-section-number">9.7.2</span> Set-up the function</a></li>
<li><a class="nav-link" href="#run-the-full-analysis"><span class="header-section-number">9.7.3</span> Run the full analysis</a></li>
<li><a class="nav-link" href="#why-is-mcmds-heritability-0"><span class="header-section-number">9.7.4</span> Why is MCMDS heritability 0?</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#de-regression-explained"><span class="header-section-number">9.8</span> De-regressed BLUPs and weighted two-stage prediction</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#computing-h2-for-weights"><span class="header-section-number">9.8.1</span> Computing H2 for weights</a></li></ul>
</li>
<li>
<a class="nav-link" href="#mixed-models-in-brief"><span class="header-section-number">9.9</span> Mixed models in brief</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#links-for-learning"><span class="header-section-number">9.9.1</span> Links for learning</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/wolfemd/GenomicSelectionManual/blob/master/05-getBLUPs.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/wolfemd/GenomicSelectionManual/edit/master/05-getBLUPs.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Genomic Prediction and Selection Manual</strong>" was written by Marnin Wolfe. It was last built on 2022-03-31.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
