<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Using this manual | Genomic Prediction and Selection Manual</title>
<meta name="author" content="Marnin Wolfe">
<meta name="description" content="3.1 Suggested workflow Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<meta name="generator" content="bookdown 0.24 with bs4_book()">
<meta property="og:title" content="Chapter 3 Using this manual | Genomic Prediction and Selection Manual">
<meta property="og:type" content="book">
<meta property="og:url" content="https://wolfemd.github.io/GenomicSelectionManual/using-this-manual.html">
<meta property="og:description" content="3.1 Suggested workflow Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Using this manual | Genomic Prediction and Selection Manual">
<meta name="twitter:description" content="3.1 Suggested workflow Create a workflowR repository for your genomic prediction analysis, following instructions here. Follow along the following documents as templates and examples: Genomic...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.11/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><link href="libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="libs/pagedtable-1.1/js/pagedtable.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="bs4_style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Genomic Prediction and Selection Manual</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Preamble</a></li>
<li><a class="" href="overview-process-map.html"><span class="header-section-number">2</span> Overview Process Map</a></li>
<li><a class="active" href="using-this-manual.html"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="" href="process-maps.html"><span class="header-section-number">4</span> Process maps</a></li>
<li><a class="" href="create_project.html"><span class="header-section-number">5</span> Create a project</a></li>
<li><a class="" href="download-training-data.html"><span class="header-section-number">6</span> Download training data</a></li>
<li><a class="" href="prepare-phenotype-data.html"><span class="header-section-number">7</span> Prepare phenotype data</a></li>
<li><a class="" href="prepare-genotypic-data.html"><span class="header-section-number">8</span> Prepare genotypic data</a></li>
<li><a class="" href="preliminary-field-trial-analysis.html"><span class="header-section-number">9</span> Preliminary field trial analysis</a></li>
<li><a class="" href="intro-to-genomic-prediction.html"><span class="header-section-number">10</span> Intro to Genomic Prediction</a></li>
<li><a class="" href="standard-k-fold-cross-validation.html"><span class="header-section-number">11</span> Standard K-fold Cross-validation</a></li>
<li><a class="" href="predict-parental-breeding-values.html"><span class="header-section-number">12</span> Predict parental breeding values</a></li>
<li><a class="" href="intro-to-genomic-cross-prediction.html"><span class="header-section-number">13</span> Intro to Genomic Cross Prediction</a></li>
<li><a class="" href="parentwise_cross_val.html"><span class="header-section-number">14</span> Accuracy of cross prediction?</a></li>
<li><a class="" href="predict-crosses.html"><span class="header-section-number">15</span> Predict crosses</a></li>
<li><a class="" href="references.html">References</a></li>
<li><a class="" href="workflowr-example.html"><span class="header-section-number">16</span> WorkflowR example</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/wolfemd/GenomicSelectionManual">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="using-this-manual" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Using this manual<a class="anchor" aria-label="anchor" href="#using-this-manual"><i class="fas fa-link"></i></a>
</h1>
<div id="suggested-workflow" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Suggested workflow<a class="anchor" aria-label="anchor" href="#suggested-workflow"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p>Create a <code>workflowR</code> repository for your genomic prediction analysis, following <a href="create_project.html#create_project">instructions here</a>.</p></li>
<li>
<p>Follow along the following documents as templates and examples:</p>
<ul>
<li><p><strong>Genomic Selection Manual (THIS BOOK)</strong></p></li>
<li>
<p><strong>GS Process Maps</strong> (see <a href="process_maps">process map section below</a>, also placed strategically throughout the manual). There are four total:</p>
<ol style="list-style-type: decimal">
<li><p>Overview Process Map</p></li>
<li><p>Data Download and Preparation Process Map</p></li>
<li><p>Preliminary Analysis and Cross-validation Process Map</p></li>
<li><p>Genomic Mate Selection Process Map</p></li>
</ol>
</li>
<li><p><strong>GS Checklist (~TO BE LINKED HERE~)</strong></p></li>
</ul>
</li>
<li><p>Use some variant of these documents and code examples to complete a genomic prediction analysis and develop a report on the results.</p></li>
<li>
<p><strong>Advice and best practices:</strong></p>
<ul>
<li><p>Choose your own data, traits, from cassavabase</p></li>
<li>
<p>Work through what the example code actually does for yourself.</p>
<ul>
<li>Follow-up functions you don’t know by going to the manual.</li>
<li>I will strive to provide references to other tutorials, papers, etc. to give context and help you learn more detail where desired..</li>
</ul>
</li>
<li><p>Inevitably, we will want divergences, alterations, bells-and-whistles on top of the process documented. <strong>SUGGEST</strong> altering and developing your own <strong>process maps</strong> and <strong>checklists</strong> as you go.</p></li>
<li><p>Use a combination of Rmarkdown (<strong>.Rmd</strong>) and Rscripts (<strong>.R</strong>) to document your analysis, as demonstrated.</p></li>
<li><p>Take the time to write commentary throughout. In full sentences, what do you intend to do? How do you interpret the results? What is the next step? Etc.</p></li>
<li><p>Take the time to think through the naming of datasets, files, folders, R objects, etc.</p></li>
<li><p>Use Git version control, made easy with Rstudio.</p></li>
<li><p>Publish your code to GitHub and a report on your results as a webpage using GitHub Pages. I will demonstrate using the package <code>workflowR</code> to manage these aspects.</p></li>
</ul>
</li>
</ul>
</div>
<div id="prerequisites" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"><i class="fas fa-link"></i></a>
</h2>
<ol style="list-style-type: decimal">
<li>You need to install R, Rstudio and relevant R packages <em>in advance</em>. Instructions are in the next section.</li>
<li>You need to know at least some R syntax. Links to learning resources are also provided in a section below. If you’ve never used R before, you are going to have trouble following the coding aspects in this manual.</li>
</ol>
</div>
<div id="install-software-and-packages" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Install software and packages<a class="anchor" aria-label="anchor" href="#install-software-and-packages"><i class="fas fa-link"></i></a>
</h2>
<div id="r-rstudio-r-packages" class="section level3" number="3.3.1">
<h3>
<span class="header-section-number">3.3.1</span> R, Rstudio, R packages<a class="anchor" aria-label="anchor" href="#r-rstudio-r-packages"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>
<p>Install R and Rstudio</p>
<ul>
<li><p><a href="https://www.datacamp.com/community/tutorials/installing-R-windows-mac-ubuntu">How to Install R on Windows, Mac OS X, and Ubuntu Tutorial</a></p></li>
<li><p><a href="https://rstudio-education.github.io/hopr/starting.html">Installing R and RStudio</a></p></li>
</ul>
</li>
<li><p><a href="https://www.rstudio.com/">Rstudio</a></p></li>
<li>
<p>Install packages</p>
<ul>
<li>
<code>tidyverse</code> (includes <code>dplyr</code>, <code>tidyr</code>, <code>ggplot2</code>, <code>magrittr</code> and other really useful ones)</li>
<li><code>genomicMateSelectR</code></li>
<li><code>workflowr</code></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tidyverse"</span>,<span class="st">"workflowr"</span>, <span class="st">"sommer"</span>, <span class="st">"lme4"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"wolfemd/genomicMateSelectR"</span>, ref <span class="op">=</span> <span class="st">'master'</span><span class="op">)</span> </code></pre></div>
</div>
<div id="create-a-github-account" class="section level3" number="3.3.2">
<h3>
<span class="header-section-number">3.3.2</span> Create a GitHub account<a class="anchor" aria-label="anchor" href="#create-a-github-account"><i class="fas fa-link"></i></a>
</h3>
<p>We would like to teach you a reproducible, open-access approach to data science and genomic selection.</p>
<p>To start, please go to <a href="https://github.com/" class="uri">https://github.com/</a> and create a free account, if you don’t already have one. We will show you how to create web-based reports, like <a href="https://wolfemd.github.io/IITA_2021GS/">this</a>, from your analyses!</p>
</div>
<div id="command-line-programs" class="section level3" number="3.3.3">
<h3>
<span class="header-section-number">3.3.3</span> Command-line Programs<a class="anchor" aria-label="anchor" href="#command-line-programs"><i class="fas fa-link"></i></a>
</h3>
<p>Using Cassavabase-derived data mostly, but not entirely removes the need for command-line informatics tools. However, I was not able to totally avoid it in the example. Furthermore, this will be a valuable skill / experience to learn.</p>
<p>For the section on <a href="prepare-genotypic-data">“preparing genotype data”</a> and further downstream when we do some steps to <a href="verify-pedigree-relationships">check the validity of the pedigree</a> we end up needing some bioinformatics tools.</p>
<div id="windows" class="section level4" number="3.3.3.1">
<h4>
<span class="header-section-number">3.3.3.1</span> Windows<a class="anchor" aria-label="anchor" href="#windows"><i class="fas fa-link"></i></a>
</h4>
<p>The three possible Linux-emulator applications that colleagues have recommended to me for Windows users:</p>
<p>1. Windows Subsystem for Linux: <a href="https://docs.microsoft.com/en-us/windows/wsl/install" class="uri">https://docs.microsoft.com/en-us/windows/wsl/install</a></p>
<p>2. Git BASH for Windows: <a href="https://gitforwindows.org/" class="uri">https://gitforwindows.org/</a></p>
<p>3. Cygwin: <a href="https://www.cygwin.com/" class="uri">https://www.cygwin.com/</a></p>
<p>I can’t give you much advice beyond those links. <strong>Get googling for solutions!</strong></p>
<p>I found this open-access google doc: <a href="http://bit.ly/2FSSjH6" class="uri">http://bit.ly/2FSSjH6</a> which <em>might</em> provide guidance for Windows users.</p>
<p><strong>BACK-UP PLAN:</strong> We might explore setting up a Cornell-based BioHPC node and allowing everyone to log-in to it remotely. The BioHPC has all the programs we could possibly want and gives access to more memory/compute cores than any single laptop.</p>
</div>
<div id="mac" class="section level4" number="3.3.3.2">
<h4>
<span class="header-section-number">3.3.3.2</span> Mac<a class="anchor" aria-label="anchor" href="#mac"><i class="fas fa-link"></i></a>
</h4>
<p>You’ll already have access to most of the commands I’ll demonstrate, e.g. <code>grep</code>, <code>cut</code>. I recommend installing “<a href="https://brew.sh/">Homebrew</a>” which will enable you to easily install e.g. <code>vcftools</code> and <code>bcftools</code> by doing e.g. <code>brew install vcftools</code> in the terminal.</p>
</div>
<div id="programs-we-might-use" class="section level4" number="3.3.3.3">
<h4>
<span class="header-section-number">3.3.3.3</span> Programs we might use<a class="anchor" aria-label="anchor" href="#programs-we-might-use"><i class="fas fa-link"></i></a>
</h4>
<ul>
<li>
<p>Bioinformatics command-line software tools:</p>
<ul>
<li><p><a href="https://vcftools.github.io/man_latest.html"><code>vcftools</code></a></p></li>
<li><p><a href="https://samtools.github.io/bcftools/bcftools.html"><code>bcftools</code></a></p></li>
<li><p><a href="https://www.cog-genomics.org/plink/"><code>plink1.9</code></a>: When it comes up late in the pipeline, I actually describe the process of getting it working on a (my) Mac laptop. In a pinch, you can download and unzip the pre-complied plink program on your machine and then use <code>./plink</code> to run the program.</p></li>
</ul>
</li>
<li><p>There are some other commands we might encounter, that should come pre-available, at least in the Mac and Linux command lines.</p></li>
</ul>
</div>
</div>
</div>
<div id="learning-r-and-more" class="section level2" number="3.4">
<h2>
<span class="header-section-number">3.4</span> Learning R and more<a class="anchor" aria-label="anchor" href="#learning-r-and-more"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><p><strong>R for Data Science (<a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a>)</strong>:<br>
As the Rstudio / Tidyverse team recommends here (<a href="https://www.tidyverse.org/learn/" class="uri">https://www.tidyverse.org/learn/</a>), this book is “the best place to start learning the tidyverse.” Just about anyone should be able to (1) read the brief introduction and (2) look over the table of contents and quickly find a starting point that meets their level / interest.</p></li>
<li>
<p><strong>Data Challenge Lab (<a href="https://datalab.stanford.edu/challenge-lab" class="uri">https://datalab.stanford.edu/challenge-lab</a>)<br></strong>Where students develop their data skills by solving a progression of increasingly difficult challenges. I just discovered this. There is much here that I think is useful. Especially amongst the “Open Content” (<a href="https://dcl-docs.stanford.edu/home/" class="uri">https://dcl-docs.stanford.edu/home/</a>). For example:</p>
<ul>
<li><p><strong>Data Wrangling</strong><br><a href="https://dcl-wrangle.stanford.edu/manip-basics.html" class="uri">https://dcl-wrangle.stanford.edu/manip-basics.html</a></p></li>
<li><p><strong>Functional Programming<br></strong><a href="https://dcl-prog.stanford.edu/">https://dcl-prog.stanford.edu/<br></a><a href="https://dcl-prog.stanford.edu/purrr-mutate.html" class="uri">https://dcl-prog.stanford.edu/purrr-mutate.html</a> (I’ve been told my loops using purrr package functions are confusing… check this out).<br>
Also:<a href="https://dcl-prog.stanford.edu/purrr-parallel.html" class="uri">https://dcl-prog.stanford.edu/purrr-parallel.html</a></p></li>
</ul>
</li>
<li><p><strong>Learn the tidyverse<br></strong><a href="https://www.tidyverse.org/learn/">https://www.tidyverse.org/learn/<br></a><a href="https://www.tidyverse.org/packages/" class="uri">https://www.tidyverse.org/packages/</a></p></li>
<li><p><strong>APS 135: Introduction to Exploratory Data Analysis with R<br></strong>(<a href="https://dzchilds.github.io/eda-for-bio/" class="uri">https://dzchilds.github.io/eda-for-bio/</a>)<br>
Has some useful intro sections on the very basics of R and Rstudio, seems to have come out of a Plant Science department, so there is that.</p></li>
<li><p><strong>Stat545: Data wrangling, exploration, and analysis with R</strong><br><a href="https://stat545.com/" class="uri">https://stat545.com/</a></p></li>
</ul>
<p><a href="https://docs.google.com/document/d/1FbT_n66WvHPL9E3hwNQJenntoS-TPgkIR-7xTBW8rBk/edit?usp=sharing"><strong>Compendium of Learning Resources gDoc</strong></a>: This google doc contains all the links/references above. It contains even more, and I hope to maintain it as a growing, dynamic, more comprehensive annotated list of resources for learning R, Rstudio, data science and more.</p>
<div id="rmarkdown-and-workflowr" class="section level3" number="3.4.1">
<h3>
<span class="header-section-number">3.4.1</span> Rmarkdown and <code>workflowR</code><a class="anchor" aria-label="anchor" href="#rmarkdown-and-workflowr"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li><a href="https://cran.r-project.org/web/packages/rmarkdown/vignettes/rmarkdown.html">Rmarkdown Vignette and Cheatsheet</a></li>
<li><a href="https://workflowr.github.io/workflowr/"><code>workflowr</code>: organized + reproducible + shareable data science in R</a></li>
</ul>
</div>
<div id="piping-code" class="section level3" number="3.4.2">
<h3>
<span class="header-section-number">3.4.2</span> Piping code <code>%&gt;%</code><a class="anchor" aria-label="anchor" href="#piping-code"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li><a href="https://www.youtube.com/watch?v=KO35_7arizY&amp;ab_channel=RsquaredAcademy">YouTube Video Intro</a></li>
<li><a href="https://magrittr.tidyverse.org/">magrittr package page</a></li>
</ul>
</div>
<div id="hotkeys" class="section level3" number="3.4.3">
<h3>
<span class="header-section-number">3.4.3</span> Hotkeys<a class="anchor" aria-label="anchor" href="#hotkeys"><i class="fas fa-link"></i></a>
</h3>
<p>Pretty critical to learn a few of these, especially these:</p>
<ul>
<li>
<code>CMD+Option+I</code> = create chunk</li>
<li>
<code>Shift+CMD+M</code> = <code>%&gt;%</code> pipe operator</li>
<li>
<code>CMD+Enter</code> = submit (run) lines of code in your Rmd or R script to the console.</li>
</ul>
</div>
</div>
<div id="r-sessions-packages-to-load" class="section level2" number="3.5">
<h2>
<span class="header-section-number">3.5</span> R sessions, packages to load<a class="anchor" aria-label="anchor" href="#r-sessions-packages-to-load"><i class="fas fa-link"></i></a>
</h2>
<p>I will use the <code>tidyverse</code> and also <code>genomicMateSelectR</code> packages throughout the pipeline.</p>
<p>There are others that may appear.</p>
<p>I recommend, for each pipeline segment, starting with a new R session. Begin each segment, with a step to load these R packages:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/genomicMateSelectR">genomicMateSelectR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gt.rstudio.com/">gt</a></span><span class="op">)</span> <span class="co"># just for the nice looking tables</span></code></pre></div>
</div>
<div id="high-performance-and-remote-computing" class="section level2" number="3.6">
<h2>
<span class="header-section-number">3.6</span> High performance and remote computing<a class="anchor" aria-label="anchor" href="#high-performance-and-remote-computing"><i class="fas fa-link"></i></a>
</h2>
<p>The example in this manual is designed to work on a laptop… at least a new one. I’ve got 16-cores and 64GB RAM on the machine I developed it on.</p>
<p>In practice, with the large number of plots, clones and SNPs that we <em>actually</em> work with, we will not use a laptop for these computations.</p>
<p>At some point, perhaps at the end of the pipeline run-through we will want to cover the (remote) use of high performance computing machines to facilitate.</p>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="overview-process-map.html"><span class="header-section-number">2</span> Overview Process Map</a></div>
<div class="next"><a href="process-maps.html"><span class="header-section-number">4</span> Process maps</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#using-this-manual"><span class="header-section-number">3</span> Using this manual</a></li>
<li><a class="nav-link" href="#suggested-workflow"><span class="header-section-number">3.1</span> Suggested workflow</a></li>
<li><a class="nav-link" href="#prerequisites"><span class="header-section-number">3.2</span> Prerequisites</a></li>
<li>
<a class="nav-link" href="#install-software-and-packages"><span class="header-section-number">3.3</span> Install software and packages</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#r-rstudio-r-packages"><span class="header-section-number">3.3.1</span> R, Rstudio, R packages</a></li>
<li><a class="nav-link" href="#create-a-github-account"><span class="header-section-number">3.3.2</span> Create a GitHub account</a></li>
<li><a class="nav-link" href="#command-line-programs"><span class="header-section-number">3.3.3</span> Command-line Programs</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#learning-r-and-more"><span class="header-section-number">3.4</span> Learning R and more</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#rmarkdown-and-workflowr"><span class="header-section-number">3.4.1</span> Rmarkdown and workflowR</a></li>
<li><a class="nav-link" href="#piping-code"><span class="header-section-number">3.4.2</span> Piping code %&gt;%</a></li>
<li><a class="nav-link" href="#hotkeys"><span class="header-section-number">3.4.3</span> Hotkeys</a></li>
</ul>
</li>
<li><a class="nav-link" href="#r-sessions-packages-to-load"><span class="header-section-number">3.5</span> R sessions, packages to load</a></li>
<li><a class="nav-link" href="#high-performance-and-remote-computing"><span class="header-section-number">3.6</span> High performance and remote computing</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/wolfemd/GenomicSelectionManual/blob/master/00-using_this_manual.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/wolfemd/GenomicSelectionManual/edit/master/00-using_this_manual.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Genomic Prediction and Selection Manual</strong>" was written by Marnin Wolfe. It was last built on 2022-03-31.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
